{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ukbb_parser import create_dataset, create_ICD10_dataset, get_chrom_raw_marker_data, get_chrom_imputation_data\n",
    "from ukbb_parser.shared_utils.util import summarize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here\n"
     ]
    }
   ],
   "source": [
    "print(\"here\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "PHENOTYPES = [\n",
    "    # The format is: ('field_name', field_id, 'field_type')\n",
    "    # field_name could be anything you like (this will determine the column name in the resulted dataframe)\n",
    "    # fiel_id is the ID of the UKBB field (e.g. http://biobank.ctsu.ox.ac.uk/crystal/field.cgi?id=50)\n",
    "    # field_type could be any of:\n",
    "    # - 'continuous' (will take the maximum among all the existing records for each sample)\n",
    "    # - 'binary' (i.e. 0 or 1)\n",
    "    # - 'set' (will return a set with all the values found for each sample; note that this is dramatically slower to parse)\n",
    "    # - You can also specify a Python function that takes all the values (as a dataframe) and returns the final values (as\n",
    "    # a series with the same index).\n",
    "    ('height', 50, 'continuous'),\n",
    "    ('diastolic_blood_preasure', 4079, 'continuous'),\n",
    "    ('red_blood_cell_count', 30010, 'continuous'),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022_05_16-12:15:30] Reading all dataset rows of 26 columns (for 8 fields)...\n",
      "[2022_05_16-12:36:15] Finished after 0:20:45.398013.\n",
      "[2022_05_16-12:36:15] Read a dataset of 502520 samples.\n",
      "[2022_05_16-12:36:15] Knowing of 141 samples who have wished to withdraw, 32 of them are in the loaded dataset. Filtering out these records, the dataset has reduced from 502520 to 502488 samples.\n",
      "[2022_05_16-12:36:16] Of 502488 samples, 472607 are self-reported whites and 409601 are Caucasians according to their genetics. Keeping only the 409569 samples that are both.\n",
      "[2022_05_16-12:36:16] Parsing field sex...\n",
      "[2022_05_16-12:36:16] Parsing field year_of_birth...\n",
      "[2022_05_16-12:36:17] To avoid the \"dummy variable trap\", removing the AC_leeds column (38017 matching records).\n",
      "[2022_05_16-12:36:21] Filtering out 0 samples without genotyping metadata.\n",
      "[2022_05_16-12:36:21] Filtering out 312 samples with mismatching genetic and self-reported sex.\n",
      "[2022_05_16-12:36:25] To avoid the \"dummy variable trap\", removing the batch_UKBiLEVEAX_b2 column (4110 matching records).\n",
      "[2022_05_16-12:36:29] Constructed 60173 kinship groups (147731 samples), of which 52417 (128259 samples) are relevant for the dataset (i.e. containing at least 2 samples in the dataset). Picking only one representative of each group and removing the 75842 other samples in those groups has reduced the dataset from 409257 to 333415 samples.\n",
      "[2022_05_16-12:36:29] Parsing field height...\n",
      "[2022_05_16-12:36:29] Parsing field diastolic_blood_preasure...\n",
      "[2022_05_16-12:36:29] Parsing field red_blood_cell_count...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    5895328\n",
       "1    5764375\n",
       "2    5503073\n",
       "3    2563626\n",
       "4    4980832\n",
       "Name: eid, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "333415 records\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>height</th>\n",
       "      <th>diastolic_blood_preasure</th>\n",
       "      <th>red_blood_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>174.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>4.780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>163.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>4.560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>184.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>4.590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>167.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>4.574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>192.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>4.725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   height  diastolic_blood_preasure  red_blood_cell_count\n",
       "0   174.0                      80.0                 4.780\n",
       "1   163.0                      94.0                 4.560\n",
       "2   184.0                      84.0                 4.590\n",
       "3   167.0                      80.0                 4.574\n",
       "4   192.0                      74.0                 4.725"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "333415 records\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>const</th>\n",
       "      <th>sex</th>\n",
       "      <th>year_of_birth</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>...</th>\n",
       "      <th>batch_UKBiLEVEAX_b1</th>\n",
       "      <th>batch_UKBiLEVEAX_b10</th>\n",
       "      <th>batch_UKBiLEVEAX_b11</th>\n",
       "      <th>batch_UKBiLEVEAX_b3</th>\n",
       "      <th>batch_UKBiLEVEAX_b4</th>\n",
       "      <th>batch_UKBiLEVEAX_b5</th>\n",
       "      <th>batch_UKBiLEVEAX_b6</th>\n",
       "      <th>batch_UKBiLEVEAX_b7</th>\n",
       "      <th>batch_UKBiLEVEAX_b8</th>\n",
       "      <th>batch_UKBiLEVEAX_b9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1942.0</td>\n",
       "      <td>-12.53270</td>\n",
       "      <td>4.23522</td>\n",
       "      <td>0.179931</td>\n",
       "      <td>-1.616470</td>\n",
       "      <td>-4.98576</td>\n",
       "      <td>-0.759869</td>\n",
       "      <td>-0.113473</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1951.0</td>\n",
       "      <td>-12.33790</td>\n",
       "      <td>3.06112</td>\n",
       "      <td>-3.053590</td>\n",
       "      <td>2.520440</td>\n",
       "      <td>6.52943</td>\n",
       "      <td>-1.429880</td>\n",
       "      <td>1.865290</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>-11.85050</td>\n",
       "      <td>5.19403</td>\n",
       "      <td>-2.353350</td>\n",
       "      <td>1.176260</td>\n",
       "      <td>3.82838</td>\n",
       "      <td>-0.501960</td>\n",
       "      <td>1.207110</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1950.0</td>\n",
       "      <td>-11.97490</td>\n",
       "      <td>3.47780</td>\n",
       "      <td>-1.490430</td>\n",
       "      <td>-0.272294</td>\n",
       "      <td>-8.48126</td>\n",
       "      <td>-4.502880</td>\n",
       "      <td>4.069820</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1948.0</td>\n",
       "      <td>-8.73549</td>\n",
       "      <td>4.98005</td>\n",
       "      <td>-1.685650</td>\n",
       "      <td>0.680946</td>\n",
       "      <td>-3.21188</td>\n",
       "      <td>-0.549570</td>\n",
       "      <td>-0.175166</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 173 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   const  sex  year_of_birth       PC1      PC2       PC3       PC4      PC5  \\\n",
       "0    1.0  1.0         1942.0 -12.53270  4.23522  0.179931 -1.616470 -4.98576   \n",
       "1    1.0  0.0         1951.0 -12.33790  3.06112 -3.053590  2.520440  6.52943   \n",
       "2    1.0  1.0         1944.0 -11.85050  5.19403 -2.353350  1.176260  3.82838   \n",
       "3    1.0  1.0         1950.0 -11.97490  3.47780 -1.490430 -0.272294 -8.48126   \n",
       "4    1.0  1.0         1948.0  -8.73549  4.98005 -1.685650  0.680946 -3.21188   \n",
       "\n",
       "        PC6       PC7  ...  batch_UKBiLEVEAX_b1  batch_UKBiLEVEAX_b10  \\\n",
       "0 -0.759869 -0.113473  ...                    0                     0   \n",
       "1 -1.429880  1.865290  ...                    0                     0   \n",
       "2 -0.501960  1.207110  ...                    0                     0   \n",
       "3 -4.502880  4.069820  ...                    0                     0   \n",
       "4 -0.549570 -0.175166  ...                    0                     0   \n",
       "\n",
       "   batch_UKBiLEVEAX_b11  batch_UKBiLEVEAX_b3  batch_UKBiLEVEAX_b4  \\\n",
       "0                     0                    0                    0   \n",
       "1                     0                    0                    0   \n",
       "2                     0                    0                    0   \n",
       "3                     0                    0                    0   \n",
       "4                     0                    0                    0   \n",
       "\n",
       "   batch_UKBiLEVEAX_b5  batch_UKBiLEVEAX_b6  batch_UKBiLEVEAX_b7  \\\n",
       "0                    0                    0                    0   \n",
       "1                    0                    0                    0   \n",
       "2                    0                    0                    0   \n",
       "3                    0                    0                    0   \n",
       "4                    0                    0                    0   \n",
       "\n",
       "   batch_UKBiLEVEAX_b8  batch_UKBiLEVEAX_b9  \n",
       "0                    0                    0  \n",
       "1                    0                    0  \n",
       "2                    0                    0  \n",
       "3                    0                    0  \n",
       "4                    0                    0  \n",
       "\n",
       "[5 rows x 173 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "333415 records\n"
     ]
    }
   ],
   "source": [
    "    # Parses the specified phenotypes. It is recommended to look at all the documentation, and in particular at all the \n",
    "# optional arguments, of the create_dataset function. In particular you should double-check that you are really\n",
    "# interested in the behavior dictated by the default values of the arguments (especially those flags which determine\n",
    "# the filtration steps).\n",
    "eid, fields, covariates = create_dataset(PHENOTYPES, nrows = None)\n",
    "\n",
    "# As a result, we get back three dataframes/series, all with the same index (so you can think of them as three pieces of\n",
    "# the same table). Each row in the table represents one sample/individual.\n",
    "\n",
    "# eid is the UKBB phenotype ID of each sample in our project (which appears in the project-specific files).\n",
    "summarize(eid)\n",
    "\n",
    "# This dataframe contains all the fields that have been explicitly requested (described by the PHENOTYPES list).\n",
    "    summarize(fields)\n",
    "\n",
    "# The covariates are variables that are always parsed and returned (unless stated otherwise in the parameters),\n",
    "# and should be accounted for if the fields are used as part of a genetic association study. These include:\n",
    "# - const: A constant 1, used as an intercept for regression/classification models.\n",
    "# - sex: 1 for males, 0 for females\n",
    "# - year_of_birth\n",
    "# - PC1-PC40: 40 principle components of PCA over the ~800K genetic markers.\n",
    "# - assessment centers (ACs) in which the samples have been recruited and processed.\n",
    "# - batches: One-hot-encoding of the batch in which the samples have been sequenced.\n",
    "summarize(covariates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2021_08_09-12:05:41] Reading all dataset rows of 345 columns (for 14 fields)...\n",
      "[2021_08_09-12:10:53] Finished after 0:05:12.543305.\n",
      "[2021_08_09-12:10:53] Read a dataset of 502520 samples.\n",
      "[2021_08_09-12:10:56] Knowing of 141 samples who have wished to withdraw, 32 of them are in the loaded dataset. Filtering out these records, the dataset has reduced from 502520 to 502488 samples.\n",
      "[2021_08_09-12:10:56] Of 502488 samples, 472607 are self-reported whites and 409601 are Caucasians according to their genetics. Keeping only the 409569 samples that are both.\n",
      "[2021_08_09-12:10:57] Parsing field sex...\n",
      "[2021_08_09-12:10:57] Parsing field year_of_birth...\n",
      "[2021_08_09-12:10:59] To avoid the \"dummy variable trap\", removing the AC_leeds column (38017 matching records).\n",
      "[2021_08_09-12:11:03] Filtering out 0 samples without genotyping metadata.\n",
      "[2021_08_09-12:11:03] Filtering out 312 samples with mismatching genetic and self-reported sex.\n",
      "[2021_08_09-12:11:04] To avoid the \"dummy variable trap\", removing the batch_UKBiLEVEAX_b2 column (4110 matching records).\n",
      "[2021_08_09-12:11:08] Constructed 60173 kinship groups (147731 samples), of which 52417 (128259 samples) are relevant for the dataset (i.e. containing at least 2 samples in the dataset). Picking only one representative of each group and removing the 75842 other samples in those groups has reduced the dataset from 409257 to 333415 samples.\n",
      "[2021_08_09-12:11:10] Parsing field height...\n",
      "[2021_08_09-12:11:10] Parsing field diastolic_blood_preasure...\n",
      "[2021_08_09-12:11:10] Parsing field red_blood_cell_count...\n",
      "[2021_08_09-12:11:16] Filtering out 58591 samples wihtout any ICD-10 codes, keeping 274824 of 333415 samples.\n",
      "[2021_08_09-12:11:18] Parsing the read dataset into an ICD-10 tree...\n",
      "[2021_08_09-12:13:14] Keeping only 12979 ICD-10 nodes that contain any samples (of 19154 nodes).\n",
      "[2021_08_09-12:13:14] Finished after 0:01:56.085790.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    5895328\n",
       "1    5764375\n",
       "2    5503073\n",
       "3    2563626\n",
       "4    4980832\n",
       "Name: eid, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274824 records\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coding</th>\n",
       "      <th>meaning</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>selectable</th>\n",
       "      <th>children_ids</th>\n",
       "      <th>samples</th>\n",
       "      <th>n_samples</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>node_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>A00</td>\n",
       "      <td>A00 Cholera</td>\n",
       "      <td>23</td>\n",
       "      <td>N</td>\n",
       "      <td>{288, 289, 287}</td>\n",
       "      <td>{379224, 341060, 349975}</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>A009</td>\n",
       "      <td>A00.9 Cholera, unspecified</td>\n",
       "      <td>286</td>\n",
       "      <td>Y</td>\n",
       "      <td>{}</td>\n",
       "      <td>{379224, 341060, 349975}</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>A01</td>\n",
       "      <td>A01 Typhoid and paratyphoid fevers</td>\n",
       "      <td>23</td>\n",
       "      <td>N</td>\n",
       "      <td>{291, 292, 293, 294, 295}</td>\n",
       "      <td>{445573, 116392, 354474, 497432, 111834}</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>A010</td>\n",
       "      <td>A01.0 Typhoid fever</td>\n",
       "      <td>290</td>\n",
       "      <td>Y</td>\n",
       "      <td>{}</td>\n",
       "      <td>{354474, 497432, 111834, 445573}</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>A014</td>\n",
       "      <td>A01.4 Paratyphoid fever, unspecified</td>\n",
       "      <td>290</td>\n",
       "      <td>Y</td>\n",
       "      <td>{}</td>\n",
       "      <td>{116392}</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        coding                               meaning  parent_id selectable  \\\n",
       "node_id                                                                      \n",
       "286        A00                           A00 Cholera         23          N   \n",
       "289       A009            A00.9 Cholera, unspecified        286          Y   \n",
       "290        A01    A01 Typhoid and paratyphoid fevers         23          N   \n",
       "291       A010                   A01.0 Typhoid fever        290          Y   \n",
       "295       A014  A01.4 Paratyphoid fever, unspecified        290          Y   \n",
       "\n",
       "                      children_ids                                   samples  \\\n",
       "node_id                                                                        \n",
       "286                {288, 289, 287}                  {379224, 341060, 349975}   \n",
       "289                             {}                  {379224, 341060, 349975}   \n",
       "290      {291, 292, 293, 294, 295}  {445573, 116392, 354474, 497432, 111834}   \n",
       "291                             {}          {354474, 497432, 111834, 445573}   \n",
       "295                             {}                                  {116392}   \n",
       "\n",
       "         n_samples  \n",
       "node_id             \n",
       "286              3  \n",
       "289              3  \n",
       "290              5  \n",
       "291              4  \n",
       "295              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12979 records\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>height</th>\n",
       "      <th>diastolic_blood_preasure</th>\n",
       "      <th>red_blood_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>174.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>4.780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>163.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>4.560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>184.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>4.590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>167.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>4.574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>192.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>4.725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   height  diastolic_blood_preasure  red_blood_cell_count\n",
       "0   174.0                      80.0                 4.780\n",
       "1   163.0                      94.0                 4.560\n",
       "2   184.0                      84.0                 4.590\n",
       "3   167.0                      80.0                 4.574\n",
       "4   192.0                      74.0                 4.725"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274824 records\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>const</th>\n",
       "      <th>sex</th>\n",
       "      <th>year_of_birth</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>...</th>\n",
       "      <th>batch_UKBiLEVEAX_b1</th>\n",
       "      <th>batch_UKBiLEVEAX_b10</th>\n",
       "      <th>batch_UKBiLEVEAX_b11</th>\n",
       "      <th>batch_UKBiLEVEAX_b3</th>\n",
       "      <th>batch_UKBiLEVEAX_b4</th>\n",
       "      <th>batch_UKBiLEVEAX_b5</th>\n",
       "      <th>batch_UKBiLEVEAX_b6</th>\n",
       "      <th>batch_UKBiLEVEAX_b7</th>\n",
       "      <th>batch_UKBiLEVEAX_b8</th>\n",
       "      <th>batch_UKBiLEVEAX_b9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1942.0</td>\n",
       "      <td>-12.53270</td>\n",
       "      <td>4.23522</td>\n",
       "      <td>0.179931</td>\n",
       "      <td>-1.616470</td>\n",
       "      <td>-4.98576</td>\n",
       "      <td>-0.759869</td>\n",
       "      <td>-0.113473</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1951.0</td>\n",
       "      <td>-12.33790</td>\n",
       "      <td>3.06112</td>\n",
       "      <td>-3.053590</td>\n",
       "      <td>2.520440</td>\n",
       "      <td>6.52943</td>\n",
       "      <td>-1.429880</td>\n",
       "      <td>1.865290</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>-11.85050</td>\n",
       "      <td>5.19403</td>\n",
       "      <td>-2.353350</td>\n",
       "      <td>1.176260</td>\n",
       "      <td>3.82838</td>\n",
       "      <td>-0.501960</td>\n",
       "      <td>1.207110</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1950.0</td>\n",
       "      <td>-11.97490</td>\n",
       "      <td>3.47780</td>\n",
       "      <td>-1.490430</td>\n",
       "      <td>-0.272294</td>\n",
       "      <td>-8.48126</td>\n",
       "      <td>-4.502880</td>\n",
       "      <td>4.069820</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1948.0</td>\n",
       "      <td>-8.73549</td>\n",
       "      <td>4.98005</td>\n",
       "      <td>-1.685650</td>\n",
       "      <td>0.680946</td>\n",
       "      <td>-3.21188</td>\n",
       "      <td>-0.549570</td>\n",
       "      <td>-0.175166</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 173 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   const  sex  year_of_birth       PC1      PC2       PC3       PC4      PC5  \\\n",
       "0    1.0  1.0         1942.0 -12.53270  4.23522  0.179931 -1.616470 -4.98576   \n",
       "1    1.0  0.0         1951.0 -12.33790  3.06112 -3.053590  2.520440  6.52943   \n",
       "2    1.0  1.0         1944.0 -11.85050  5.19403 -2.353350  1.176260  3.82838   \n",
       "3    1.0  1.0         1950.0 -11.97490  3.47780 -1.490430 -0.272294 -8.48126   \n",
       "4    1.0  1.0         1948.0  -8.73549  4.98005 -1.685650  0.680946 -3.21188   \n",
       "\n",
       "        PC6       PC7  ...  batch_UKBiLEVEAX_b1  batch_UKBiLEVEAX_b10  \\\n",
       "0 -0.759869 -0.113473  ...                    0                     0   \n",
       "1 -1.429880  1.865290  ...                    0                     0   \n",
       "2 -0.501960  1.207110  ...                    0                     0   \n",
       "3 -4.502880  4.069820  ...                    0                     0   \n",
       "4 -0.549570 -0.175166  ...                    0                     0   \n",
       "\n",
       "   batch_UKBiLEVEAX_b11  batch_UKBiLEVEAX_b3  batch_UKBiLEVEAX_b4  \\\n",
       "0                     0                    0                    0   \n",
       "1                     0                    0                    0   \n",
       "2                     0                    0                    0   \n",
       "3                     0                    0                    0   \n",
       "4                     0                    0                    0   \n",
       "\n",
       "   batch_UKBiLEVEAX_b5  batch_UKBiLEVEAX_b6  batch_UKBiLEVEAX_b7  \\\n",
       "0                    0                    0                    0   \n",
       "1                    0                    0                    0   \n",
       "2                    0                    0                    0   \n",
       "3                    0                    0                    0   \n",
       "4                    0                    0                    0   \n",
       "\n",
       "   batch_UKBiLEVEAX_b8  batch_UKBiLEVEAX_b9  \n",
       "0                    0                    0  \n",
       "1                    0                    0  \n",
       "2                    0                    0  \n",
       "3                    0                    0  \n",
       "4                    0                    0  \n",
       "\n",
       "[5 rows x 173 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274824 records\n"
     ]
    }
   ],
   "source": [
    "# The function create_ICD10_dataset is similar to create_dataset (with mostly overlapping arguments). The main difference\n",
    "# between the two functions is that create_ICD10_dataset also returns an ICD10_tree for the samples. The ICD10_tree is\n",
    "# yet another dataframe, with each row representing a node of the ICD-10 tree. The 'samples' column provides the set of\n",
    "# all sample indices associated with each ICD-10 code (these indices are simply the index of the other three dataframes). \n",
    "eid, ICD10_tree, additional_fields, covariates, _ = create_ICD10_dataset(PHENOTYPES, nrows = None)\n",
    "summarize(eid)\n",
    "summarize(ICD10_tree)\n",
    "summarize(additional_fields)\n",
    "summarize(covariates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examples working with the phenotypic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'additional_fields' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-9cf7773bc5dd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Example 1: A graph showing diastolic blood preasure vs. red blood cell count\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscatter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madditional_fields\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'diastolic_blood_preasure'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0madditional_fields\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'red_blood_cell_count'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'additional_fields' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD8CAYAAAB0IB+mAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Z1A+gAAAACXBIWXMAAAsTAAALEwEAmpwYAAANT0lEQVR4nO3cYYjkd33H8ffHO1NpjKb0VpC706T00njYQtIlTRFqirZc8uDugUXuIFgleGAbKVWEFEuU+MiGWhCu1ZOKVdAYfSALntwDjQTEC7chNXgXItvTeheFrDHNk6Ax7bcPZtKdrneZf3Zndy/7fb/gYP7/+e3Mlx97752d2ZlUFZKk7e8VWz2AJGlzGHxJasLgS1ITBl+SmjD4ktSEwZekJqYGP8lnkzyZ5PuXuD5JPplkKcmjSW6c/ZiSpPUa8gj/c8CBF7n+VmDf+N9R4F/WP5YkadamBr+qHgR+/iJLDgGfr5FTwNVJXj+rASVJs7FzBrexGzg/cXxhfO6nqxcmOcrotwCuvPLKP7z++utncPeS1MfDDz/8s6qaW8vXziL4g1XVceA4wPz8fC0uLm7m3UvSy16S/1zr187ir3SeAPZOHO8Zn5MkXUZmEfwF4F3jv9a5GXimqn7t6RxJ0taa+pROki8BtwC7klwAPgK8EqCqPgWcAG4DloBngfds1LCSpLWbGvyqOjLl+gL+emYTSZI2hO+0laQmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqYlBwU9yIMnjSZaS3HWR69+Q5IEkjyR5NMltsx9VkrQeU4OfZAdwDLgV2A8cSbJ/1bK/B+6vqhuAw8A/z3pQSdL6DHmEfxOwVFXnquo54D7g0Ko1BbxmfPm1wE9mN6IkaRaGBH83cH7i+ML43KSPArcnuQCcAN5/sRtKcjTJYpLF5eXlNYwrSVqrWb1oewT4XFXtAW4DvpDk1267qo5X1XxVzc/Nzc3oriVJQwwJ/hPA3onjPeNzk+4A7geoqu8CrwJ2zWJASdJsDAn+aWBfkmuTXMHoRdmFVWt+DLwNIMmbGAXf52wk6TIyNfhV9TxwJ3ASeIzRX+OcSXJPkoPjZR8E3pvke8CXgHdXVW3U0JKkl27nkEVVdYLRi7GT5+6euHwWeMtsR5MkzZLvtJWkJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNTEo+EkOJHk8yVKSuy6x5p1JziY5k+SLsx1TkrReO6ctSLIDOAb8GXABOJ1koarOTqzZB/wd8JaqejrJ6zZqYEnS2gx5hH8TsFRV56rqOeA+4NCqNe8FjlXV0wBV9eRsx5QkrdeQ4O8Gzk8cXxifm3QdcF2S7yQ5leTAxW4oydEki0kWl5eX1zaxJGlNZvWi7U5gH3ALcAT4TJKrVy+qquNVNV9V83NzczO6a0nSEEOC/wSwd+J4z/jcpAvAQlX9qqp+CPyA0Q8ASdJlYkjwTwP7klyb5ArgMLCwas3XGD26J8kuRk/xnJvdmJKk9Zoa/Kp6HrgTOAk8BtxfVWeS3JPk4HjZSeCpJGeBB4APVdVTGzW0JOmlS1VtyR3Pz8/X4uLilty3JL1cJXm4qubX8rW+01aSmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmBgU/yYEkjydZSnLXi6x7R5JKMj+7ESVJszA1+El2AMeAW4H9wJEk+y+y7irgb4CHZj2kJGn9hjzCvwlYqqpzVfUccB9w6CLrPgZ8HPjFDOeTJM3IkODvBs5PHF8Yn/s/SW4E9lbV11/shpIcTbKYZHF5efklDytJWrt1v2ib5BXAJ4APTltbVcerar6q5ufm5tZ715Kkl2BI8J8A9k4c7xmfe8FVwJuBbyf5EXAzsOALt5J0eRkS/NPAviTXJrkCOAwsvHBlVT1TVbuq6pqqugY4BRysqsUNmViStCZTg19VzwN3AieBx4D7q+pMknuSHNzoASVJs7FzyKKqOgGcWHXu7kusvWX9Y0mSZs132kpSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmhgU/CQHkjyeZCnJXRe5/gNJziZ5NMk3k7xx9qNKktZjavCT7ACOAbcC+4EjSfavWvYIMF9VfwB8FfiHWQ8qSVqfIY/wbwKWqupcVT0H3AccmlxQVQ9U1bPjw1PAntmOKUlaryHB3w2cnzi+MD53KXcA37jYFUmOJllMsri8vDx8SknSus30RdsktwPzwL0Xu76qjlfVfFXNz83NzfKuJUlT7Byw5glg78TxnvG5/yfJ24EPA2+tql/OZjxJ0qwMeYR/GtiX5NokVwCHgYXJBUluAD4NHKyqJ2c/piRpvaYGv6qeB+4ETgKPAfdX1Zkk9yQ5OF52L/Bq4CtJ/j3JwiVuTpK0RYY8pUNVnQBOrDp398Tlt894LknSjPlOW0lqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpoYFPwkB5I8nmQpyV0Xuf43knx5fP1DSa6Z+aSSpHWZGvwkO4BjwK3AfuBIkv2rlt0BPF1Vvwv8E/DxWQ8qSVqfIY/wbwKWqupcVT0H3AccWrXmEPBv48tfBd6WJLMbU5K0XjsHrNkNnJ84vgD80aXWVNXzSZ4Bfhv42eSiJEeBo+PDXyb5/lqG3oZ2sWqvGnMvVrgXK9yLFb+31i8cEvyZqarjwHGAJItVNb+Z93+5ci9WuBcr3IsV7sWKJItr/dohT+k8AeydON4zPnfRNUl2Aq8FnlrrUJKk2RsS/NPAviTXJrkCOAwsrFqzAPzl+PJfAN+qqprdmJKk9Zr6lM74Ofk7gZPADuCzVXUmyT3AYlUtAP8KfCHJEvBzRj8Upjm+jrm3G/dihXuxwr1Y4V6sWPNexAfiktSD77SVpCYMviQ1seHB92MZVgzYiw8kOZvk0STfTPLGrZhzM0zbi4l170hSSbbtn+QN2Ysk7xx/b5xJ8sXNnnGzDPg/8oYkDyR5ZPz/5LatmHOjJflskicv9V6ljHxyvE+PJrlx0A1X1Yb9Y/Qi738AvwNcAXwP2L9qzV8BnxpfPgx8eSNn2qp/A/fiT4HfHF9+X+e9GK+7CngQOAXMb/XcW/h9sQ94BPit8fHrtnruLdyL48D7xpf3Az/a6rk3aC/+BLgR+P4lrr8N+AYQ4GbgoSG3u9GP8P1YhhVT96KqHqiqZ8eHpxi952E7GvJ9AfAxRp/L9IvNHG6TDdmL9wLHquppgKp6cpNn3CxD9qKA14wvvxb4ySbOt2mq6kFGf/F4KYeAz9fIKeDqJK+fdrsbHfyLfSzD7kutqarngRc+lmG7GbIXk+5g9BN8O5q6F+NfUfdW1dc3c7AtMOT74jrguiTfSXIqyYFNm25zDdmLjwK3J7kAnADevzmjXXZeak+ATf5oBQ2T5HZgHnjrVs+yFZK8AvgE8O4tHuVysZPR0zq3MPqt78Ekv19V/7WVQ22RI8Dnquofk/wxo/f/vLmq/merB3s52OhH+H4sw4ohe0GStwMfBg5W1S83abbNNm0vrgLeDHw7yY8YPUe5sE1fuB3yfXEBWKiqX1XVD4EfMPoBsN0M2Ys7gPsBquq7wKsYfbBaN4N6stpGB9+PZVgxdS+S3AB8mlHst+vztDBlL6rqmaraVVXXVNU1jF7POFhVa/7QqMvYkP8jX2P06J4kuxg9xXNuE2fcLEP24sfA2wCSvIlR8Jc3dcrLwwLwrvFf69wMPFNVP532RRv6lE5t3McyvOwM3It7gVcDXxm/bv3jqjq4ZUNvkIF70cLAvTgJ/HmSs8B/Ax+qqm33W/DAvfgg8Jkkf8voBdx3b8cHiEm+xOiH/K7x6xUfAV4JUFWfYvT6xW3AEvAs8J5Bt7sN90qSdBG+01aSmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElq4n8BzPZculjwdYoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example 1: A graph showing diastolic blood preasure vs. red blood cell count\n",
    "fig, ax = plt.subplots()\n",
    "ax.scatter(additional_fields['diastolic_blood_preasure'], additional_fields['red_blood_cell_count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Text(1, 0, 'F'), Text(2, 0, 'M')]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Z1A+gAAAACXBIWXMAAAsTAAALEwEAmpwYAAAWnklEQVR4nO3df2zU933H8efbZ4OLWQserEohga6B7sKtaxMvZZunxmk3wpYUIm1rXKkk5RSrWfCyLBWFnrR0f1jqWH6osbUiJlshWriSbVmKumSsC6dFJw0mU7WJG28DraExSxt3JFlwZjg77/3hr63DMdi+8/lrf+71kE73vff3e+YNnF58+dzn+/mauyMiImGpibsBERGZewp3EZEAKdxFRAKkcBcRCZDCXUQkQLVxNwCwatUqX79+fdxtiIgsKidPnvyZu6+eat+CCPf169fT29sbdxsiIouKmZ253D4Ny4iIBEjhLiISIIW7iEiAFO4iIgFSuIuIBEjhLiIVl81mSaVSJBIJUqkU2Ww27paCtyCmQopIuLLZLJlMhu7ubpqbm8nn86TTaQBaW1tj7i5cthCW/G1qanLNcxcJUyqVorOzk5aWlolaLpejvb2dvr6+GDtb/MzspLs3TblP4S4ilZRIJBgeHqaurm6iVigUqK+vZ3R0NMbOFr8rhbvG3EWkopLJJPl8/pJaPp8nmUzG1FF10Ji7iFRUJpPhc5/7HA0NDZw5c4Z169YxNDTEN77xjbhbC5rCXUQq7u2332ZwcBCAV155hfr6+pg7Cp+GZUSkonbt2kWhUODhhx9maGiIhx9+mEKhwK5du+JuLWj6QlVEKsrM2LBhA6dPn8bdMTOuvfZaTp06xULIn8XsSl+oalhGRCru1KlTE9vufslrqQwNy4iIBEjhLiISIIW7iEiApg13M7vazHJm9rKZ/dDM7ovqjWb2XTM7FT2vjOpmZo+Z2Wkze9HMrq/0b0JEFofHH3887haqxkzO3EeAB9z9OmAzcK+ZXQfsAZ539w3A89FrgK3AhujRBnxzzrsWkUXprrvuiruFqjFtuLv7a+7+vWj7baAfWANsAw5Ghx0Etkfb24AnfMxxYIWZXTXXjYuIyOXNaszdzNYDnwBOAB9099eiXT8BPhhtrwFeLXrbQFQTEZF5MuNwN7PlwN8Bf+zu/1u8z8euRJjV1Qhm1mZmvWbWO35ZsoiIzI0ZhbuZ1TEW7E+6+9NR+afjwy3R8+tR/SxwddHb10a1S7j7AXdvcvem1atXl9q/iIhMYSazZQzoBvrd/ZGiXUeAO6PtO4FvF9V3RLNmNgNvFQ3fiIjIPJjJ8gO/AXwBeMnMvh/Vvgp8HXjKzNLAGeAPon3PAr8DnAbeAb44lw2LiMj0ZjJbJu/u5u4fc/ePR49n3f1/3P3T7r7B3T/j7uei493d73X3j7j7L7u7VgSbJ7oJsYiM0xWqgchms9x3330MDQ0BMDQ0xH333aeAF6lSCvdA7N69m9raWnp6ehgeHqanp4fa2lp2794dd2siEgOFeyAGBgY4ePAgLS0t1NXV0dLSwsGDBxkYGIi7NRGJgcJdRCRACvdArF27lh07dpDL5SgUCuRyOXbs2MHatWvjbk2qkJlNPGZ63HTHyuwo3AOxb98+hoaG2LJlC0uWLGHLli0MDQ2xb9++uFuTKuTuE4+ZHqdb7s0thXtAJp/56ExIFoLL3QhbN8iuLIV7IHbv3s2yZcs4evQoFy9e5OjRoyxbtkyzZSR2nZ2d7Nq1i6VLlwKwdOlSdu3aRWdnZ8ydhc0Wwn+FmpqavLdX1zqVw8z47Gc/y9GjR7lw4QJLly5ly5YtHDlyRP/dlQXDzPR5nENmdtLdm6bctxD+oBXu5bvSEMxC+DsWAYX7XLtSuGtYJjDLly+/5FlEqpPCPTDnz5+/5FlEqpPCXUQkQAp3EZEAKdxFRAKkcBcRCZDCXUQkQAp3EZEAKdxFRAKkcBcRCZDCXUQkQNOGu5n1mNnrZtZXVPu4mR03s++bWa+Z3RjVzcweM7PTZvaimV1fyeZFRGRqMzlzfxy4ZVJtH/Bn7v5x4E+j1wBbgQ3Row345px0KSIiszJtuLv7C8C5yWXg/dH2B4D/jra3AU/4mOPACjO7aq6alUuVciszEakOtSW+74+Bo2b2EGP/QPx6VF8DvFp03EBUe23yDzCzNsbO7rnmmmtKbKO6FS+dqiV/JQ6NjY288cYbs3rPbE8yVq5cyblzk88vZTqlfqF6D3C/u18N3A90z/YHuPsBd29y96bVq1eX2IaIxOmNN954z31Q5/ox2388ZEyp4X4n8HS0/TfAjdH2WeDqouPWRjWpsMudneusXaQ6lRru/w18Ktq+GTgVbR8BdkSzZjYDb7n7e4ZkpDKK7yCvu8mLVLdpx9zNLAvcBKwyswHgQeBu4BtmVgsME42dA88CvwOcBt4BvliBnkVEZBrThru7t15m1w1THOvAveU2JSIi5Sl1toyICP7g++FrH6j8ryGzpnAXkZLZn/1vxb/bMTP8axX9JYKktWVERAKkcBcRCZDCXUQkQAp3EZEAKdxFRAKk2TIiUpZKrza6cuXKiv78UCncRaRks50GaWZaFmOeaFhGRCRACncRkQAp3EVEAqRwFxEJkMJ9gWtsbLzkHqgzeQCzOr6xsTHm36WIzDXNllngxm9jVkm6cbZIeHTmLiISIIW7iEiAFO4iIgFSuIuIBEjhLiISoGlny5hZD3Ar8Lq7p4rq7YzdDHsU+Ad33x3V9wLpqP5H7n60Eo1XC92jUkRKMZOpkI8DXcAT4wUzawG2Ab/i7hfM7Bei+nXAHcAm4EPAP5vZRncfnevGq4XuUSkipZh2WMbdXwDOTSrfA3zd3S9Ex7we1bcB33L3C+7+I+A0cOMc9isiIjNQ6pj7RuA3zeyEmf2Lmf1qVF8DvFp03EBUExGReVTqFaq1QCOwGfhV4Ckz+8XZ/AAzawPaAK655poS2xARkamUeuY+ADztY/4NeBdYBZwFri46bm1Uew93P+DuTe7etHr16hLbEJHFYKp1j6SySg33Z4AWADPbCCwBfgYcAe4ws6Vm9mFgA/Bvc9BnVZvtwmGzfeg2ZlJJlwtyBXxlzWQqZBa4CVhlZgPAg0AP0GNmfcBF4E4fm9LxQzN7CngZGAHu1UyZ8pQyU0a3MpO4zTS4Jx+nz+3csYXwh9nU1OS9vb1xtxEMhbssJFcKen1Oy2NmJ929aap9ukJVRCRACncRkQAp3EVEAqRwFxEJkMJdRCRACncRkQAp3EVEAqRwF5F5c//998fdQtVQuIvIvHn00UfjbqFqKNxFRAKkcBcRCZDCXUQkQAp3EZEAKdxFZF7U19dz/Phx6uvr426lKpR6mz0RkVkZHh5m8+bNcbdRNRTuASleN3t8W+tli1QnDcsEQrcyk4Vu+fLllzxLZenMfREr5VZmOpOXOJgZ58+fB+D8+fO6W9g80Jn7IubuE4+5OE6kUtydRCIBQCKR0GdxHijcA/SVr3wl7hZE3qOtrY0333yTtra2uFupCrpBdiB0E2JZqMyM+vp6RkdHKRQK1NXVkUgkGB4e1mezTGXdINvMeszsdTPrm2LfA2bmZrYqem1m9piZnTazF83s+vLbF5HF7vbbb2fjxo3U1NSwceNGbr/99rhbCt5MhmUeB26ZXDSzq4HfBn5cVN4KbIgebcA3y29RRBazxsZGDh8+zM6dO3n77bfZuXMnhw8fprGxMe7WgjZtuLv7C8C5KXY9CuwGiv9ftQ14wsccB1aY2VVz0qmILEpdXV0sX76cPXv20NDQwJ49e1i+fDldXV1xtxa0kr5QNbNtwFl3/8GkXWuAV4teD0S1qX5Gm5n1mlnv4OBgKW2IyCLQ2trK/v37LxmW2b9/P62trXG3FrRZz3M3s2XAVxkbkimZux8ADsDYF6rl/CwRWdhaW1sV5vOslIuYPgJ8GPhBNENjLfA9M7sROAtcXXTs2qgmIiLzaNbDMu7+krv/gruvd/f1jA29XO/uPwGOADuiWTObgbfc/bW5bVlERKYzk6mQWeBfgY+a2YCZpa9w+LPAfwGngb8C/nBOuhQRkVmZdljG3a84UBadvY9vO3Bv+W2JiEg5tPyAiEiAFO4iUnHZbJZUKkUikSCVSpHNZuNuKXha8jcwdXV1E+t3FAqFuNsRIZvNkslk6O7uprm5mXw+Tzo99tWdpkdWjhYOC0RtbS2jo6PvqScSCUZGRmLoSGRMKpVi+/btPPPMM/T395NMJide9/W9Z8kqmYUrLRymM/dAjI6OYmbU1NQwOjpKIpHg3XffnTLwRebTyy+/zNDQED09PRNn7jt37uTMmTNxtxY0jbkH5LbbbmNkZAR3Z2RkhNtuuy3ulkRYsmQJ7e3ttLS0UFdXR0tLC+3t7SxZsiTu1oKmM/eAHDlyhLq6OkZGRqitrdVwjCwIFy9epKuri0984hMTZ+5dXV1cvHgx7taCpnAPRCKRYHR0dOLmB+PP47c2E4nLddddx/bt22lvb58Yc//85z/PM888E3drQdOwTCDcnZqaGvbt28fQ0BD79u2jpqZGd7qR2GUyGQ4dOkRnZyfDw8N0dnZy6NAhMplM3K0FTbNlAmFm3HzzzeRyOdwdM6OlpYVjx44p4CV22WyWjo6OiTP3TCajaZBzoKzb7MniUFtby4kTJ1i3bh01NTWsW7eOEydOUFurkTeRaqRwD8TSpUsZGhpi69atnDt3jq1btzI0NMTSpUvjbk2q3PhFTMXDMplMRlepVpiGZQJhZqxfv55XXnllojb+eiH8HUv1SqVSdHZ20tLSMlHL5XK0t7frIqYyXWlYRuEeiOjGKRNTIIunQi6Ev2OpXolEguHhYerq6iZqhUKB+vp6XWRXJo25V5G7776bN998k7vvvjvuVkQASCaT5PP5S2r5fJ5kMhlTR9VB4R6QhoYGnnvuOVauXMlzzz1HQ0ND3C2JkMlkSKfT5HI5CoUCuVyOdDqtqZAVpqkUAbn11lvp6+vDzGhoaOCTn/wkhw8fjrstqXLjUx6LL2Lq6OjQVMgK05h7IGpra3n33Xd56KGH+NKXvsT+/fv58pe/TE1NjZYhEAmUxtyrwD333IO788ADD9DQ0MADDzyAu3PPPffE3ZqIxEDhHpDxGTOXey0i1WPacDezHjN73cz6imp/YWb/bmYvmtnfm9mKon17zey0mf2HmW2pUN8yyf79+1mxYgXHjh3j4sWLHDt2jBUrVrB///64WxORGMzkzP1x4JZJte8CKXf/GPCfwF4AM7sOuAPYFL3nL81MyxLOg5GREZ588slL1sx+8sknNd4uUqWmDXd3fwE4N6n2T+4+nhrHgbXR9jbgW+5+wd1/BJwGbpzDfuUKJl/tp6v/RKrXXIy57wSei7bXAK8W7RuIalJhjY2N7N27l0ceeYR33nmHRx55hL1799LY2Bh3ayJks1lSqRSJRIJUKqV1ZeZBWeFuZhlgBHiyhPe2mVmvmfUODg6W04YAXV1dLFu2jD179tDQ0MCePXtYtmwZXV1dcbcmVU4Lh8XE3ad9AOuBvkm1u4B/BZYV1fYCe4teHwV+bbqff8MNN7iU79ChQ75p0yavqanxTZs2+aFDh+JuScQ3bdrkx44du6R27Ngx37RpU0wdhQPo9cvk6owuYjKz9cB33D0Vvb4FeAT4lLsPFh23CTjE2Dj7h4DngQ3ufsXVgXQRk0i4tHBY5ZR1EZOZZRk7Q/+omQ2YWRroAn4O+K6Zfd/M9gO4+w+Bp4CXgX8E7p0u2EUkbFo4LB7Tri3j7lMtANF9heM7gI5ymhKRcIwvHNbd3U1zczP5fJ50Ok1Hh2KikrRwmIhUlBYOi4cWDhMRWaS0cJiISJVRuIuIBEjhLiISIIW7iEiAFO4iIgFSuIuIBEjhLiISIIW7iEiAFO4iIgFSuIuIBEjhLiISIIW7iEiAFO4iIgFSuIuIBEjhLiISIIW7iEiAFO4iIgFSuIuIBEjhLiISoGnD3cx6zOx1M+srqjWa2XfN7FT0vDKqm5k9ZmanzexFM7u+ks3LpbLZLKlUikQiQSqVIpvNxt2SiMRkJmfujwO3TKrtAZ539w3A89FrgK3AhujRBnxzbtqU6WSzWTKZDJ2dnQwPD9PZ2Ukmk1HAi1SpacPd3V8Azk0qbwMORtsHge1F9Sd8zHFghZldNUe9yhV0dHTQ3d1NS0sLdXV1tLS00N3dTUdHR9ytiUgMSh1z/6C7vxZt/wT4YLS9Bni16LiBqPYeZtZmZr1m1js4OFhiGzKuv7+f5ubmS2rNzc309/fH1JGIxKnsL1Td3QEv4X0H3L3J3ZtWr15dbhtVL5lMks/nL6nl83mSyWRMHYlInEoN95+OD7dEz69H9bPA1UXHrY1qUmGZTIZ0Ok0ul6NQKJDL5Uin02QymbhbE5EY1Jb4viPAncDXo+dvF9V3mdm3gE8CbxUN30gFtba2AtDe3k5/fz/JZJKOjo6JuohUFxsbVbnCAWZZ4CZgFfBT4EHgGeAp4BrgDPAH7n7OzAzoYmx2zTvAF929d7ommpqavLd32sNERKSImZ1096ap9k175u7ulzv1+/QUxzpw7+zaExGRuaYrVEVEAqRwFxEJkMJdRCRACncRkQAp3EVEAqRwFxEJkMJdRCRACncRkQAp3EVEAqRwFxEJkMJdRCRACncRkQAp3EVEAqRwFxEJkMJdRCRACncRkQAp3EVEAqRwFxEJkMJdRCRACncRkQCVFe5mdr+Z/dDM+swsa2b1ZvZhMzthZqfN7LCZLZmrZkVEZGZKDnczWwP8EdDk7ikgAdwB/DnwqLtfC7wBpOeiURERmblyh2VqgfeZWS2wDHgNuBn422j/QWB7mb+GiIjMUsnh7u5ngYeAHzMW6m8BJ4E33X0kOmwAWFNukyIiMjvlDMusBLYBHwY+BDQAt8zi/W1m1mtmvYODg6W2ISIiUyhnWOYzwI/cfdDdC8DTwG8AK6JhGoC1wNmp3uzuB9y9yd2bVq9eXUYbIiIyWTnh/mNgs5ktMzMDPg28DOSA34uOuRP4dnktiojIbJUz5n6CsS9Ovwe8FP2sA8BXgD8xs9PAzwPdc9CniIjMQlmzZdz9QXf/JXdPufsX3P2Cu/+Xu9/o7te6+++7+4W5alZEFqdsNksqlSKRSJBKpchms3G3FLza6Q8RESldNpslk8nQ3d1Nc3Mz+XyedHrs8pfW1taYuwuXuXvcPdDU1OS9vb1xtyEiFZBKpejs7KSlpWWilsvlaG9vp6+vL8bOFj8zO+nuTVPuU7iLSCUlEgmGh4epq6ubqBUKBerr6xkdHY2xs8XvSuGuhcNEpKKSyST5fP6SWj6fJ5lMxtRRdVC4i0hFZTIZ0uk0uVyOQqFALpcjnU6TyWTibi1o+kJVRCpq/EvT9vZ2+vv7SSaTdHR06MvUCtOYu4jIIqUxdxGRKqNwFxEJkMJdRCRACncRkQAp3EVEArQgZsuY2SBwJu4+ArIK+FncTYhMQZ/NubXO3ae8IcaCCHeZW2bWe7npUSJx0mdz/mhYRkQkQAp3EZEAKdzDdCDuBkQuQ5/NeaIxdxGRAOnMXUQkQAp3EZEAKdwDYmajZvb9osf6uHsSMTM3s78uel1rZoNm9p04+wqd1nMPy/+5+8fjbkJkkiEgZWbvc/f/A34LOBtzT8HTmbuIzIdngd+NtluBbIy9VAWFe1jeVzQk8/dxNyNS5FvAHWZWD3wMOBFzP8HTsExYNCwjC5K7vxh9B9TK2Fm8VJjCXUTmyxHgIeAm4OfjbSV8CncRmS89wJvu/pKZ3RRzL8FTuIvIvHD3AeCxuPuoFlp+QEQkQJotIyISIIW7iEiAFO4iIgFSuIuIBEjhLiISIIW7iEiAFO4iIgH6f34UTnsmQ1DwAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example 2: A graph showing height distribution per sex\n",
    "fig, ax = plt.subplots()\n",
    "ax.boxplot([additional_fields.loc[covariates['sex'] == 0, 'height'].dropna(), \\\n",
    "        additional_fields.loc[covariates['sex'] == 1, 'height'].dropna()])\n",
    "ax.set_xticklabels(['F', 'M'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Working with genetic data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Raw markers (BED files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2021_08_09-12:13:33] Created link: /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_snp_chr17_v2.bim -> /tmp/ukb_roeizucker_chr17.bim\n",
      "[2021_08_09-12:13:33] Created link: /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_cal_chr17_v2.bed -> /tmp/ukb_roeizucker_chr17.bed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:06<00:00,  2.28s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17</td>\n",
       "      <td>rs2294076</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6157</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>rs2396789</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8547</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>rs36068254</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10583</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17</td>\n",
       "      <td>rs6565705</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13905</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17</td>\n",
       "      <td>rs2294075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14008</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom         snp   cm    pos a0 a1  i\n",
       "0    17   rs2294076  0.0   6157  A  G  0\n",
       "1    17   rs2396789  0.0   8547  T  C  1\n",
       "2    17  rs36068254  0.0  10583  A  G  2\n",
       "3    17   rs6565705  0.0  13905  G  A  3\n",
       "4    17   rs2294075  0.0  14008  G  C  4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28835 total variants.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fid</th>\n",
       "      <th>iid</th>\n",
       "      <th>father</th>\n",
       "      <th>mother</th>\n",
       "      <th>gender</th>\n",
       "      <th>trait</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1688461</td>\n",
       "      <td>1688461</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5444187</td>\n",
       "      <td>5444187</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2676900</td>\n",
       "      <td>2676900</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3709342</td>\n",
       "      <td>3709342</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3769417</td>\n",
       "      <td>3769417</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fid      iid father mother gender trait  i\n",
       "0  1688461  1688461      0      0      1  <NA>  0\n",
       "1  5444187  5444187      0      0      2  <NA>  1\n",
       "2  2676900  2676900      0      0      2  <NA>  2\n",
       "3  3709342  3709342      0      0      2  <NA>  3\n",
       "4  3769417  3769417      0      0      2  <NA>  4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "488377 records\n"
     ]
    }
   ],
   "source": [
    "# Load the genetic data for chr17.\n",
    "bim, fam, G = get_chrom_raw_marker_data('17')\n",
    "\n",
    "# 'bim' is a dataframe containing information about the variants.\n",
    "display(bim.head())\n",
    "print('%d total variants.' % len(bim))\n",
    "\n",
    "# 'fam' is a dataframe containing information about the samples (specifically, we need it to map the sample IDs).\n",
    "assert (fam['i'] == np.arange(len(fam))).all()\n",
    "summarize(fam)\n",
    "\n",
    "# G is the genotyping matrix (see usage below)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "488377\n"
     ]
    }
   ],
   "source": [
    "# print(G)\n",
    "print(len(G[5].compute()))\n",
    "# relevant_variants = bim[bim['pos'].between(59500000, 59590000)].copy()\n",
    "# for i, (_, variant) in enumerate(relevant_variants.iterrows()):\n",
    "#     print(variant)\n",
    "#     variant_genotypes = G[variant['i'], :].compute()\n",
    "#     print((variant_genotypes))\n",
    "    \n",
    "#     print('Variant %d/%d' % (i + 1, len(relevant_variants)), end = '\\r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    197572\n",
       "1    151419\n",
       "2    222977\n",
       "3    213019\n",
       "4    279357\n",
       "Name: eid, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274824 records\n"
     ]
    }
   ],
   "source": [
    "# In order to query the genotypes corresponding to samples in the phenotypic dataset we've parsed, we create a series\n",
    "# mapping each sample_index (i.e. the index shared by 'eid', 'additional_fields', 'covariates' etc.) to its correspondig \n",
    "# index within the G matrix. This is done by creating a series mapping between the 'iid' (= eid) to the 'i' columns in the\n",
    "# 'fam' dataframe. Note that the column 'i' provides the index within G of each sample.\n",
    "genotyping_index = eid.map(fam.astype({'iid': int}).set_index('iid')['i'])\n",
    "summarize(genotyping_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant 5/5\r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "      <th>height_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20242</th>\n",
       "      <td>17</td>\n",
       "      <td>rs4968433</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59500333</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>20242</td>\n",
       "      <td>1.114311e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20243</th>\n",
       "      <td>17</td>\n",
       "      <td>rs878745</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59502319</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>20243</td>\n",
       "      <td>1.193244e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20244</th>\n",
       "      <td>17</td>\n",
       "      <td>rs9896616</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59516472</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>20244</td>\n",
       "      <td>3.137731e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20245</th>\n",
       "      <td>17</td>\n",
       "      <td>rs62071328</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59517034</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>20245</td>\n",
       "      <td>1.194888e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20246</th>\n",
       "      <td>17</td>\n",
       "      <td>rs146944614</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59518015</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>20246</td>\n",
       "      <td>2.970597e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom          snp   cm       pos a0 a1      i   height_pval\n",
       "20242    17    rs4968433  0.0  59500333  G  A  20242  1.114311e-15\n",
       "20243    17     rs878745  0.0  59502319  T  C  20243  1.193244e-22\n",
       "20244    17    rs9896616  0.0  59516472  T  C  20244  3.137731e-09\n",
       "20245    17   rs62071328  0.0  59517034  T  C  20245  1.194888e-01\n",
       "20246    17  rs146944614  0.0  59518015  G  A  20246  2.970597e-02"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import statsmodels.api as sm\n",
    "\n",
    "# As a simple example, we will run association tests for each of the the variants in the gonomic locus\n",
    "# chr17:59,500,000-59,520,000 against the height phenotype.\n",
    "relevant_variants = bim[bim['pos'].between(59500000, 59520000)].copy()\n",
    "\n",
    "pval_results = []\n",
    "\n",
    "for i, (_, variant) in enumerate(relevant_variants.iterrows()):\n",
    "    \n",
    "    print('Variant %d/%d' % (i + 1, len(relevant_variants)), end = '\\r')\n",
    "    \n",
    "    # variant['i'] gives the index of the variant, so taking this row in the G matrix will give the genotypes of all the\n",
    "    # UKBB samples for this variant. The values in the G matrix indicate the genotype of each individual for each variant.\n",
    "    # These values can be 0 (homozygous reference, i.e. no copies of the alternative allele), 1 (heterozygous, i.e. one\n",
    "    # copy of the alternative allele) or 2 (homozygous alternative, i.e. two copies of the alternative allele).\n",
    "    variant_genotypes = G[variant['i'], :].compute()\n",
    "    # We use the genotyping_index to retrieve only the relevant samples from the loaded genotypes (and in the right order),\n",
    "    # so we end up with a vector of genotypes aligned to our dataset.\n",
    "    sample_variant_genotypes = variant_genotypes[genotyping_index.values]\n",
    "    sample_variant_genotypes = pd.Series(sample_variant_genotypes, index = eid.index).rename('genotype')\n",
    "    \n",
    "    # We construct the relevant dataset for this variant. The exogenous variables (X) will be all the covariates, along\n",
    "    # with the genotypes (the variable we are really interested in) and the endogenous variable (y) will be the height.\n",
    "    mask = pd.notnull(sample_variant_genotypes) & pd.notnull(additional_fields['height'])\n",
    "    y = additional_fields.loc[mask, 'height']\n",
    "    X = pd.concat([covariates, sample_variant_genotypes], axis = 1).loc[mask]\n",
    "    \n",
    "    # We train a simple linear-regression model (OLS) to calculate a p-value for this variant's genotypes (as a variable\n",
    "    # in the OLS, after accounting for all other covariates).\n",
    "    model = sm.OLS(y, X)\n",
    "    variant_results = model.fit()\n",
    "    pval = variant_results.pvalues['genotype']\n",
    "    pval_results.append(pval)\n",
    "\n",
    "# We finalize and display the results (the p-value derived for any of the variants).\n",
    "relevant_variants['height_pval'] = pval_results\n",
    "display(relevant_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation genetic data (BGEN files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>rsid</th>\n",
       "      <th>number_of_alleles</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>file_start_position</th>\n",
       "      <th>size_in_bytes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>60086</td>\n",
       "      <td>rs537499872</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>15609040</td>\n",
       "      <td>52348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>60105</td>\n",
       "      <td>rs556230663</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>15661388</td>\n",
       "      <td>7586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>60112</td>\n",
       "      <td>rs575737384</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>15668974</td>\n",
       "      <td>5894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16</td>\n",
       "      <td>60164</td>\n",
       "      <td>rs181534180</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>15674868</td>\n",
       "      <td>110212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>60169</td>\n",
       "      <td>rs558005370</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>15785080</td>\n",
       "      <td>12271</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome  position         rsid  number_of_alleles allele1 allele2  \\\n",
       "0         16     60086  rs537499872                  2       G       C   \n",
       "1         16     60105  rs556230663                  2       G       C   \n",
       "2         16     60112  rs575737384                  2       G       A   \n",
       "3         16     60164  rs181534180                  2       G       T   \n",
       "4         16     60169  rs558005370                  2       G       A   \n",
       "\n",
       "   file_start_position  size_in_bytes  \n",
       "0             15609040          52348  \n",
       "1             15661388           7586  \n",
       "2             15668974           5894  \n",
       "3             15674868         110212  \n",
       "4             15785080          12271  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3089612 total variants.\n",
      "487409 samples with imputed variants on this chromosome.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    1688461\n",
       "1    5444187\n",
       "2    2676900\n",
       "3    3709342\n",
       "4    3769417\n",
       "Name: eid, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "487409 records\n"
     ]
    }
   ],
   "source": [
    "# We load the imputation data for chr16.\n",
    "chrom_imputation_reader = get_chrom_imputation_data('16')\n",
    "\n",
    "# All the imputed variants on the chromosome can be accessed through this dataframe. \n",
    "display(chrom_imputation_reader.variants.head())\n",
    "\n",
    "print('%d total variants.' % chrom_imputation_reader.n_variants)\n",
    "print('%d samples with imputed variants on this chromosome.' % chrom_imputation_reader.n_samples)\n",
    "\n",
    "# Provides the eid corresponding to each sample in the imputation reader.\n",
    "summarize(chrom_imputation_reader.sample_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274215 of the 274824 dataset samples have imputated variants on this chromosome.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    197344\n",
       "1    151251\n",
       "2    222712\n",
       "3    212765\n",
       "4    278931\n",
       "Name: eid, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274215 records\n"
     ]
    }
   ],
   "source": [
    "# In order to query the imputed genotypes corresponding to samples in the phenotypic dataset we've parsed, we create a\n",
    "# series mapping each sample_index (i.e. the index shared by 'eid', 'additional_fields', 'covariates' etc.) to its\n",
    "# correspondig index within chrom_imputation_reader. This is done by creating a series mapping between the 'sample_ids'\n",
    "# (= eid) to the index.\n",
    "imputation_index = eid.map(chrom_imputation_reader.sample_ids.reset_index().set_index('eid')['index'])\n",
    "imputation_mask = pd.notnull(imputation_index)\n",
    "print('%d of the %d dataset samples have imputated variants on this chromosome.' % (imputation_mask.sum(), \\\n",
    "        len(imputation_mask)))\n",
    "imputation_index = imputation_index[imputation_mask].astype(int)\n",
    "summarize(imputation_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chromosome                      16\n",
       "position                  53830465\n",
       "rsid                     rs9930506\n",
       "number_of_alleles                2\n",
       "allele1                          A\n",
       "allele2                          G\n",
       "file_start_position    43160279967\n",
       "size_in_bytes               135527\n",
       "Name: 1583331, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 186256 carriers of the variant.\n",
      "12414 cases, 261801 controls.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cases</th>\n",
       "      <th>controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>carriers</th>\n",
       "      <td>8817</td>\n",
       "      <td>177439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>non_carriers</th>\n",
       "      <td>3597</td>\n",
       "      <td>84362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              cases  controls\n",
       "carriers       8817    177439\n",
       "non_carriers   3597     84362"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fisher's exact test results: OR = 1.17, p-value = 2.2e-14\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import fisher_exact\n",
    "\n",
    "'''\n",
    "As a simple example, we consider the association between the variant rs9930506 in the FTO gene on chr16 to obesity (ICD-10\n",
    "code E66). According to SNPedia, there should be an association between them (https://www.snpedia.com/index.php/Rs9930506).\n",
    "'''\n",
    "\n",
    "# We query the record of the variant we look for (given by rsid).\n",
    "(_, variant_record), = chrom_imputation_reader.variants[chrom_imputation_reader.variants['rsid'] == 'rs9930506'].iterrows()\n",
    "display(variant_record)\n",
    "\n",
    "# We read the genotyping for the variant by providing the variant index (which is the name of the relevant row in the\n",
    "# chrom_imputation_reader.variants dataframe).\n",
    "variant_genotyping_probs = chrom_imputation_reader.read_variant_probs(variant_record.name)\n",
    "# We use imputation_index to retrieve only the relevant samples from the loaded genotypes (and in the right order),\n",
    "# so we end up with a vector of genotypes aligned to our dataset.\n",
    "sample_variant_genotyping_probs = variant_genotyping_probs[imputation_index.values, :]\n",
    "# The variant's genotyping is given as an Nx3 matrix, where N is the number of samples. Each row in the matrix\n",
    "# represents the genotyping of a specific individual. Since imputation is probabalistic, the genotyping is described as\n",
    "# three probabilities for having 0, 1 or 2 copies of the alternative allele, respectively (these are the same options\n",
    "# for the raw markers in the BED files, as described above).\n",
    "# For our purpose, we define a sample to be a carrier of the variant if there's at least 50% chance of it having any copy\n",
    "# of the alternative allele (under an assumed dominant effect), i.e. if the probability for 0 copies is smaller than 0.5.\n",
    "carrier_mask = (sample_variant_genotyping_probs[:, 0] <= 0.5)\n",
    "print('There are %d carriers of the variant.' % carrier_mask.sum())\n",
    "\n",
    "# We extract the node with ICD-10 coding E66, which stands for obesity.\n",
    "(_, ICD10_node), = ICD10_tree[ICD10_tree['coding'] == 'E66'].iterrows()\n",
    "# We create a mask vector, aligned to the samples used in imputation data, that would determine whether each sample is\n",
    "# inflicted with obesity.\n",
    "case_mask = pd.Series([(sample_index in ICD10_node['samples']) for sample_index in imputation_index.index], index = \\\n",
    "        imputation_index.index)\n",
    "control_mask = ~case_mask\n",
    "print('%d cases, %d controls.' % (case_mask.sum(), control_mask.sum()))\n",
    "\n",
    "# We construct a contingency table counting the number of sample with each combination of being either a carrier or\n",
    "# non-carrier of the allele, and being either of the case or the control group.\n",
    "contingency_table = pd.DataFrame([\n",
    "    [(carrier_mask & case_mask).sum(), (carrier_mask & control_mask).sum()],\n",
    "    [((~carrier_mask) & case_mask).sum(), ((~carrier_mask) & control_mask).sum()],\n",
    "], index = ['carriers', 'non_carriers'], columns = ['cases', 'controls'])\n",
    "display(contingency_table)\n",
    "\n",
    "# We run Fisher's exact test to test the reported association.\n",
    "print('Fisher\\'s exact test results: OR = %.2f, p-value = %.2g' % fisher_exact(contingency_table))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tip:\n",
    "\n",
    "It is recommended to subscribe to the UKB-GENETICS@JISCMAIL.AC.UK mailing list (https://www.jiscmail.ac.uk/cgi-bin/webadmin?A0=UKB-GENETICS), where you can see the questions, answers and discussion of other researchers working with the UKBB genetic data, and also post your own questions regarding the dataset."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
