{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a2.3LM 64-bit Intel (24 Jan 2020)\r\n"
     ]
    }
   ],
   "source": [
    "!/cs/phd/nadavb/third_party/plink2 --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/cs/labs/michall/roeizucker/plink_results/cancer/results’: File exists\n",
      "mkdir: cannot create directory ‘/cs/labs/michall/roeizucker/plink_results/cancer/phenotypes’: File exists\n",
      "PLINK v2.00a2.3LM 64-bit Intel (24 Jan 2020)   www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Error: Failed to open /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer/a.txt.log for logging: No such file or directory.\n"
     ]
    }
   ],
   "source": [
    "!mkdir /cs/labs/michall/roeizucker/plink_results/cancer/results\n",
    "!mkdir /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes\n",
    "!touch /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer.log\n",
    "!touch /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/pan_cancer.txt\n",
    "!touch /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt\n",
    "!/cs/phd/nadavb/third_party/plink2 --bed /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_cal_chrY_v2.bed --bim /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_snp_chrY_v2.bim --fam /cs/phd/nadavb/big_data/ukbiobank/genetics/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/pan_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer/ --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 1 --memory 10000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a2.3LM 64-bit Intel (24 Jan 2020)   www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer1.log.\n",
      "Options in effect:\n",
      "  --1\n",
      "  --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt\n",
      "  --covar-variance-standardize\n",
      "  --glm hide-covar\n",
      "  --mac 20\n",
      "  --memory 10000\n",
      "  --out /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer1\n",
      "  --pgen /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_cal_chrY_v2.bed\n",
      "  --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/pan_cancer.txt\n",
      "  --psam /cs/phd/nadavb/big_data/ukbiobank/genetics/ukb2666_cal_chr1_v2_s488366.fam\n",
      "  --pvar /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_snp_chrY_v2.bim\n",
      "  --threads 1\n",
      "\n",
      "Start time: Tue Aug 31 19:48:44 2021\n",
      "161081 MiB RAM detected; reserving 10000 MiB for main workspace.\n",
      "Using 1 compute thread.\n",
      "488377 samples (264808 females, 223474 males, 95 ambiguous; 488377 founders)\n",
      "loaded from\n",
      "/cs/phd/nadavb/big_data/ukbiobank/genetics/ukb2666_cal_chr1_v2_s488366.fam.\n",
      "691 variants loaded from\n",
      "/cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_snp_chrY_v2.bim.\n",
      "2 phenotypes loaded (1 binary, 1 categorical).\n",
      "172 covariates loaded from /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt.\n",
      "--covar-variance-standardize: 172 covariates transformed.\n",
      "Calculating allele frequencies... done.\n",
      "412 variants removed due to allele frequency threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "279 variants remaining after main filters.\n",
      "--glm: Skipping categorical phenotype 'PHENO1'.\n",
      "Warning: Not including covariate 'COVAR1' in --glm regression on phenotype\n",
      "'PHENO2'.\n",
      "--glm logistic regression on phenotype 'PHENO2': done.\n",
      "Results written to /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer1.PHENO2.glm.logistic .\n",
      "End time: Tue Aug 31 20:01:18 2021\n"
     ]
    }
   ],
   "source": [
    "!/cs/phd/nadavb/third_party/plink2 --bed /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_cal_chrY_v2.bed --bim /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_snp_chrY_v2.bim --fam /cs/phd/nadavb/big_data/ukbiobank/genetics/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/pan_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer1 --1 --glm hide-covar --mac 20 --covar-variance-standardize  --threads 1 --memory 10000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import json\n",
    "from datetime import datetime\n",
    "from collections import Counter\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/cs/phd/nadavb/pwas_results/cancer_phenotype_columns.json'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-68dfdbd6248b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/cs/phd/nadavb/pwas_results/cancer_phenotype_columns.json'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mcancer_phenotype_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/cs/phd/nadavb/pwas_results/ukbb_covariate_columns.json'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mcovariate_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/cs/phd/nadavb/pwas_results/cancer_phenotype_columns.json'"
     ]
    }
   ],
   "source": [
    "with open('/cs/phd/nadavb/pwas_results/cancer_phenotype_columns.json', 'r') as f:\n",
    "    cancer_phenotype_cols = json.load(f)\n",
    "    \n",
    "with open('/cs/phd/nadavb/pwas_results/ukbb_covariate_columns.json', 'r') as f:\n",
    "    covariate_cols = json.load(f)\n",
    "\n",
    "cancer_dataset = pd.read_csv('/cs/phd/nadavb/pwas_results/ukbb_cancer_dataset.csv')\n",
    "\n",
    "# print(cancer_phenotype_cols)\n",
    "# print(covariate_cols)\n",
    "# display(cancer_dataset)\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Breast cancer\n",
      "Epithelial ovarian cancer\n",
      "Prostate cancer\n",
      "Colorectal cancer\n",
      "Lung cancer\n",
      "Chronic lymphocytic leukemia\n",
      "Pancreatic cancer\n",
      "Melanoma\n",
      "Non-melanoma skin cancer\n",
      "Pan-Cancer\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "for cancer_phenotype_col in cancer_phenotype_cols:\n",
    "    print(cancer_phenotype_col)\n",
    "    cancer_file_name = cancer_phenotype_col.lower().replace(' ', '_').replace('-', '_') + '.txt'\n",
    "    cancer_values = cancer_dataset[['eid', 'eid', cancer_phenotype_col]].dropna()\n",
    "    cancer_values.to_csv(os.path.join('/cs/labs/michall/roeizucker/plink_results/cancer/phenotypes', cancer_file_name), \\\n",
    "            header = False, index = False, sep = '\\t')\n",
    "\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cancer_dataset' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-f98302dd4191>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m covariate_values = cancer_dataset[['eid', 'eid'] + [covariate_col for covariate_col in covariate_cols if \\\n\u001b[0m\u001b[1;32m      2\u001b[0m         covariate_col != 'const']].dropna()\n\u001b[1;32m      3\u001b[0m \u001b[0mcovariate_values\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Done.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cancer_dataset' is not defined"
     ]
    }
   ],
   "source": [
    "covariate_values = cancer_dataset[['eid', 'eid'] + [covariate_col for covariate_col in covariate_cols if \\\n",
    "        covariate_col != 'const']].dropna()\n",
    "covariate_values.to_csv('/cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt', header = False, index = False, sep = '\\t')\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test a PLINK command on chrY"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "rm -f /cs/phd/nadavb/plink_results/cancer/results/*\n",
    "/cs/phd/nadavb/third_party/plink2 --bed /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_cal_chrY_v2.bed --bim /cs/phd/nadavb/big_data/ukbiobank/genetics/EGAD00010001226/001/ukb_snp_chrY_v2.bim --fam /cs/phd/nadavb/big_data/ukbiobank/genetics/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/phd/nadavb/plink_results/cancer/phenotypes/pan_cancer.txt --covar /cs/phd/nadavb/plink_results/cancer/covariates.txt --out /cs/phd/nadavb/plink_results/cancer/results/pan_cancer --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 1 --memory 10000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQS</th>\n",
       "      <th>OBS_CT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y</td>\n",
       "      <td>rs1800865</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>24934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89009265</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>223034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89005343</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>223234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89024591</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999973</td>\n",
       "      <td>223318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89014410</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999987</td>\n",
       "      <td>222571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89022554</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>223121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89013956</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>223330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89024729</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>222935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89022046</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>223154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>Y</td>\n",
       "      <td>Affx-89007087</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.998184</td>\n",
       "      <td>223006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>691 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    #CHROM             ID REF ALT  ALT_FREQS  OBS_CT\n",
       "0        Y      rs1800865   A   G   1.000000   24934\n",
       "1        Y  Affx-89009265   G   T   0.999996  223034\n",
       "2        Y  Affx-89005343   G   A   0.999996  223234\n",
       "3        Y  Affx-89024591   A   G   0.999973  223318\n",
       "4        Y  Affx-89014410   A   G   0.999987  222571\n",
       "..     ...            ...  ..  ..        ...     ...\n",
       "686      Y  Affx-89022554   C   G   1.000000  223121\n",
       "687      Y  Affx-89013956   G   A   1.000000  223330\n",
       "688      Y  Affx-89024729   G   A   1.000000  222935\n",
       "689      Y  Affx-89022046   T   A   0.999996  223154\n",
       "690      Y  Affx-89007087   C   T   0.998184  223006\n",
       "\n",
       "[691 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "afreqs = pd.read_csv('/cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer.afreq', sep = '\\t')\n",
    "display(afreqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>OBS_CT</th>\n",
       "      <th>AF</th>\n",
       "      <th>OR</th>\n",
       "      <th>LOG(OR)_SE</th>\n",
       "      <th>Z_STAT</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y</td>\n",
       "      <td>2734854</td>\n",
       "      <td>rs35284970</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125507</td>\n",
       "      <td>0.001653</td>\n",
       "      <td>0.699136</td>\n",
       "      <td>0.495417</td>\n",
       "      <td>-0.722441</td>\n",
       "      <td>0.470023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>2756471</td>\n",
       "      <td>rs9785784</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125490</td>\n",
       "      <td>0.045911</td>\n",
       "      <td>1.013280</td>\n",
       "      <td>0.044109</td>\n",
       "      <td>0.299078</td>\n",
       "      <td>0.764880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y</td>\n",
       "      <td>2828425</td>\n",
       "      <td>rs2267802</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125230</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y</td>\n",
       "      <td>2846401</td>\n",
       "      <td>Affx-89021408</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125198</td>\n",
       "      <td>0.018255</td>\n",
       "      <td>0.945296</td>\n",
       "      <td>0.055620</td>\n",
       "      <td>-1.011460</td>\n",
       "      <td>0.311796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y</td>\n",
       "      <td>2866994</td>\n",
       "      <td>Affx-89024406</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125410</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>1.092840</td>\n",
       "      <td>0.423471</td>\n",
       "      <td>0.209638</td>\n",
       "      <td>0.833950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Y</td>\n",
       "      <td>23244026</td>\n",
       "      <td>Affx-52321335</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125427</td>\n",
       "      <td>0.018095</td>\n",
       "      <td>0.943111</td>\n",
       "      <td>0.054114</td>\n",
       "      <td>-1.082380</td>\n",
       "      <td>0.279085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>Y</td>\n",
       "      <td>23582147</td>\n",
       "      <td>Affx-89006643</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125393</td>\n",
       "      <td>0.001600</td>\n",
       "      <td>1.084940</td>\n",
       "      <td>0.166880</td>\n",
       "      <td>0.488504</td>\n",
       "      <td>0.625193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>Y</td>\n",
       "      <td>23631793</td>\n",
       "      <td>Affx-89016629</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125382</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>1.028560</td>\n",
       "      <td>0.294867</td>\n",
       "      <td>0.095516</td>\n",
       "      <td>0.923905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Y</td>\n",
       "      <td>23632251</td>\n",
       "      <td>Affx-89019908</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125385</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.728866</td>\n",
       "      <td>0.398079</td>\n",
       "      <td>-0.794480</td>\n",
       "      <td>0.426916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>Y</td>\n",
       "      <td>28767492</td>\n",
       "      <td>Affx-89007087</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>125363</td>\n",
       "      <td>0.001816</td>\n",
       "      <td>1.008700</td>\n",
       "      <td>0.162335</td>\n",
       "      <td>0.053390</td>\n",
       "      <td>0.957421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>279 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    #CHROM       POS             ID REF ALT A1 TEST  OBS_CT        AF  \\\n",
       "0        Y   2734854     rs35284970   T   C  T  ADD  125507  0.001653   \n",
       "1        Y   2756471      rs9785784   A   T  A  ADD  125490  0.045911   \n",
       "2        Y   2828425      rs2267802   G   A  G  ADD  125230  0.000175   \n",
       "3        Y   2846401  Affx-89021408   A   C  A  ADD  125198  0.018255   \n",
       "4        Y   2866994  Affx-89024406   C   G  C  ADD  125410  0.000238   \n",
       "..     ...       ...            ...  ..  .. ..  ...     ...       ...   \n",
       "274      Y  23244026  Affx-52321335   T   C  T  ADD  125427  0.018095   \n",
       "275      Y  23582147  Affx-89006643   C   G  C  ADD  125393  0.001600   \n",
       "276      Y  23631793  Affx-89016629   A   G  A  ADD  125382  0.000480   \n",
       "277      Y  23632251  Affx-89019908   T   C  T  ADD  125385  0.000327   \n",
       "278      Y  28767492  Affx-89007087   C   T  C  ADD  125363  0.001816   \n",
       "\n",
       "           OR  LOG(OR)_SE    Z_STAT         P  \n",
       "0    0.699136    0.495417 -0.722441  0.470023  \n",
       "1    1.013280    0.044109  0.299078  0.764880  \n",
       "2         NaN         NaN       NaN       NaN  \n",
       "3    0.945296    0.055620 -1.011460  0.311796  \n",
       "4    1.092840    0.423471  0.209638  0.833950  \n",
       "..        ...         ...       ...       ...  \n",
       "274  0.943111    0.054114 -1.082380  0.279085  \n",
       "275  1.084940    0.166880  0.488504  0.625193  \n",
       "276  1.028560    0.294867  0.095516  0.923905  \n",
       "277  0.728866    0.398079 -0.794480  0.426916  \n",
       "278  1.008700    0.162335  0.053390  0.957421  \n",
       "\n",
       "[279 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "id_to_AF = 1 - afreqs.set_index('ID')['ALT_FREQS']\n",
    "plink_results = pd.read_csv('/cs/labs/michall/roeizucker/plink_results/cancer/results/pan_cancer1.PHENO2.glm.logistic', sep = '\\t')\n",
    "plink_results_AF = plink_results['ID'].map(id_to_AF)\n",
    "plink_results.insert(plink_results.columns.tolist().index('OR'), 'AF', plink_results_AF)\n",
    "display(plink_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running everything on the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submission script is ready: /cs/labs/michall/roeizucker/plink_results/cancer/submit_jobs.sh\n"
     ]
    }
   ],
   "source": [
    "# Original parameters:                          --threads 1 --memory 10000 and -c1 --mem=12g\n",
    "# First re-run parameters (with RERUN = True):  --threads 8 --memory 30000 and -c8 --mem=36g\n",
    "# Second re-run parameters (with RERUN = True): --threads 32 --memory 100000 and -c32 --mem=128g\n",
    "\n",
    "RERUN = False\n",
    "\n",
    "PLINK_CMD_TEMPLATE = '/cs/phd/nadavb/third_party/plink2 --bed %s --bim %s --fam %s --pheno ' + \\\n",
    "        '%s --covar %s --out %s --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq ' + \\\n",
    "        '--threads 32 --memory 100000'\n",
    "SLURM_CMD_TEMPLATE = 'sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/%s --mem=128g -c32 --time=14-0 --wrap=\"%s\"'\n",
    "SH_SCRIPT_FIRST_LINE = '#! /bin/tcsh'\n",
    "\n",
    "OUTPUT_SUBMISSION_SCRIPT_FILE_PATH = '/cs/labs/michall/roeizucker/plink_results/cancer/submit_jobs.sh'\n",
    "LOG_FILE_NAME_TEMPLATE = 'cancer_2020_04_20_%s.out'\n",
    "COVAR_FILE_PATH = '/cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt'\n",
    "PHENOTYPES_DIR = '/cs/labs/michall/roeizucker/plink_results/cancer/phenotypes'\n",
    "RESULTS_DIR = '/cs/labs/michall/roeizucker/plink_results/cancer/results'\n",
    "GENETIC_DATA_DIR = '/cs/icore/idomarg/covarathon/001'\n",
    "\n",
    "genotyping_spec = pd.read_csv('/cs/phd/nadavb/pwas_results/ukbb_raw_marker_genotyping_spec.csv')\n",
    "\n",
    "def convert_genetic_file_path(original_path):\n",
    "    updated_path = os.path.join(GENETIC_DATA_DIR, os.path.basename(original_path))\n",
    "    assert os.path.exists(updated_path)\n",
    "    return updated_path\n",
    "\n",
    "with open(OUTPUT_SUBMISSION_SCRIPT_FILE_PATH, 'w') as f:\n",
    "    \n",
    "    f.write(SH_SCRIPT_FIRST_LINE + '\\n\\n')\n",
    "    \n",
    "    for cancer_phenotype_col in cancer_phenotype_cols:\n",
    "\n",
    "        cancer_simplified_name = cancer_phenotype_col.lower().replace(' ', '_').replace('-', '_')\n",
    "        cancer_phenotype_file_path = os.path.join(PHENOTYPES_DIR, '%s.txt' % cancer_simplified_name)\n",
    "\n",
    "        for _, (chr_name, bed_file_path, bim_file_path, fam_file_path) in genotyping_spec[['name', 'bed_file_path', \\\n",
    "                'bim_file_path', 'fam_file_path']].iterrows():\n",
    "            \n",
    "            if RERUN:\n",
    "                if (results_status['is_done'] & (results_status['chrom'] == chr_name) & \\\n",
    "                        (results_status['phenotype_name'] == cancer_simplified_name)).sum() > 0:\n",
    "                    continue\n",
    "\n",
    "            bed_file_path, bim_file_path, fam_file_path = map(convert_genetic_file_path, [bed_file_path, bim_file_path, \\\n",
    "                    fam_file_path])\n",
    "            task_name = 'chr%s_%s' % (chr_name, cancer_simplified_name)\n",
    "            output_path_prefix = os.path.join(RESULTS_DIR, task_name)\n",
    "            plink_cmd = PLINK_CMD_TEMPLATE % (bed_file_path, bim_file_path, fam_file_path, cancer_phenotype_file_path, \\\n",
    "                    COVAR_FILE_PATH, output_path_prefix)\n",
    "            log_file_name = LOG_FILE_NAME_TEMPLATE % task_name\n",
    "            slurm_cmd = SLURM_CMD_TEMPLATE % (log_file_name, plink_cmd)\n",
    "            f.write(slurm_cmd + '\\n')\n",
    "            \n",
    "print('Submission script is ready: %s' % OUTPUT_SUBMISSION_SCRIPT_FILE_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#! /bin/tcsh\r\n",
      "\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr1_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr1_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr1_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr1_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr2_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr2_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr2_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr2_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr3_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr3_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr3_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr3_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr4_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr4_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr4_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr4_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr5_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr5_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr5_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr5_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr6_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr6_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr6_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr6_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr7_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr7_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr7_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr7_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n",
      "sbatch --output=/cs/labs/michall/roeizucker/plink_results/logs/slurm_jobs/cancer_2020_04_20_chr8_breast_cancer.out --mem=128g -c32 --time=14-0 --wrap=\"/cs/phd/nadavb/third_party/plink2 --bed /cs/icore/idomarg/covarathon/001/ukb_cal_chr8_v2.bed --bim /cs/icore/idomarg/covarathon/001/ukb_snp_chr8_v2.bim --fam /cs/icore/idomarg/covarathon/001/ukb2666_cal_chr1_v2_s488366.fam --pheno /cs/labs/michall/roeizucker/plink_results/cancer/phenotypes/breast_cancer.txt --covar /cs/labs/michall/roeizucker/plink_results/cancer/covariates.txt --out /cs/labs/michall/roeizucker/plink_results/cancer/results/chr8_breast_cancer --1 --glm hide-covar no-x-sex --mac 20 --covar-variance-standardize --freq --threads 32 --memory 100000\"\r\n"
     ]
    }
   ],
   "source": [
    "!head /cs/labs/michall/roeizucker/plink_results/cancer/submit_jobs.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validating that it's all finished"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Viewing the log files to see which jobs have successfully finished.\n",
    "\n",
    "RESULTS_DIR = '/cs/phd/nadavb/plink_results/cancer/results'\n",
    "\n",
    "LOG_FILE_NAME_PATTERN = re.compile(r'chr([\\dXYMT]+)_(\\S+)\\.log')\n",
    "\n",
    "log_file_paths = [os.path.join(RESULTS_DIR, file_name) for file_name in os.listdir(RESULTS_DIR) if \\\n",
    "        file_name.endswith('.log')]\n",
    "\n",
    "results_status = []\n",
    "\n",
    "for log_file_path in log_file_paths:\n",
    "    \n",
    "    (chrom, phenotype_name), = LOG_FILE_NAME_PATTERN.findall(os.path.basename(log_file_path))\n",
    "    \n",
    "    with open(log_file_path, 'r') as f:\n",
    "        is_done = 'done.' in f.read()\n",
    "        \n",
    "    results_status.append((phenotype_name, chrom, is_done, log_file_path))\n",
    "    \n",
    "results_status = pd.DataFrame(results_status, columns = ['phenotype_name', 'chrom', 'is_done', 'log_file'])\\\n",
    "        .sort_values(['phenotype_name', 'chrom'])\n",
    "display(results_status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "existing_combinatios = [(result['phenotype_name'], result['chrom']) for _, result in results_status.iterrows()]\n",
    "missing_combinatios = 0\n",
    "\n",
    "for phenotype_name in sorted(results_status['phenotype_name'].unique()):\n",
    "    for chrom in sorted(results_status['chrom'].unique()):\n",
    "        \n",
    "        combination = (phenotype_name, chrom)\n",
    "        \n",
    "        if combination not in existing_combinatios:\n",
    "            print(combination)\n",
    "            missing_combinatios += 1\n",
    "            \n",
    "print('There are %d missing log files.' % missing_combinatios)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_status['is_done'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_status.loc[~results_status['is_done']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_status.loc[~results_status['is_done'], 'log_file'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collecting the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXPECTED_MISSING_RESULTS = {\n",
    "    # cancer_simplified_name, chr_name\n",
    "    ('breast_cancer', 'Y'),\n",
    "    ('epithelial_ovarian_cancer', 'Y'),\n",
    "}\n",
    "\n",
    "COMBINED_RESULTS_DIR = '/cs/phd/nadavb/plink_results/cancer/combined_results'\n",
    "\n",
    "for cancer_phenotype_col in cancer_phenotype_cols:\n",
    "    \n",
    "    cancer_simplified_name = cancer_phenotype_col.lower().replace(' ', '_').replace('-', '_')\n",
    "    cancer_gwas_results = []\n",
    "    \n",
    "    for chr_name in genotyping_spec['name']:\n",
    "        \n",
    "        if (cancer_simplified_name, chr_name) in EXPECTED_MISSING_RESULTS:\n",
    "            continue\n",
    "        \n",
    "        result_path_prefix = os.path.join(RESULTS_DIR, 'chr%s_%s' % (chr_name, cancer_simplified_name))\n",
    "        \n",
    "        chrom_gwas_results = pd.read_csv(result_path_prefix + '.PHENO2.glm.logistic', sep = '\\t')\\\n",
    "                .rename(columns = {'#CHROM': 'CHROM'})\n",
    "        chrom_AF_results = pd.read_csv(result_path_prefix + '.afreq', sep = '\\t').rename(columns = {'#CHROM': 'CHROM'})\n",
    "        snp_id_to_af = 1 - chrom_AF_results.set_index('ID')['ALT_FREQS']\n",
    "        chrom_gwas_results['AF'] = chrom_gwas_results['ID'].map(snp_id_to_af)\n",
    "        cancer_gwas_results.append(chrom_gwas_results)\n",
    "        \n",
    "    cancer_gwas_results = pd.concat(cancer_gwas_results).reset_index(drop = True)\n",
    "    print('%s: %d variant results.' % (cancer_simplified_name, len(cancer_gwas_results)))\n",
    "    cancer_gwas_results.to_csv(os.path.join(COMBINED_RESULTS_DIR, cancer_simplified_name + '.csv'), index = False)\n",
    "\n",
    "print('Done.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
