{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "926b1159",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f12f45e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e9461571",
   "metadata": {},
   "outputs": [],
   "source": [
    "# script consts\n",
    "PHEN_START = 2\n",
    "VALUE_AT_END_OF_PHENOTYPES = 'const'\n",
    "PLINK_PATH = \"/sci/nosnap/michall/roeizucker/plink2\"\n",
    "BATCH_LENGTH = 250\n",
    "MAX_JOBS_BEFORE_SUBMITTING_NEW_ARRAY = 200\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "714996ba-5d72-4194-9b30-ef307fdee293",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # input files\n",
    "# PHEN_PATH = \"/sci/nosnap/michall/roeizucker/Hypothyroidism_hackaton/ukbb_dataset_filtered.csv\"\n",
    "# GWAS_PATH = \"/sci/nosnap/michall/roeizucker/Hypothyroidism_hackaton/GWAS_FOR_PRS\"\n",
    "# PHENOTYPES = [\"E03_F_reduced\",\"E03_M_reduced\",\"E03_reduced\",\n",
    "#              \"E04_F_reduced\",\"E04_M_reduced\",\"E04_reduced\",\n",
    "#              \"E05_F_reduced\",\"E05_M_reduced\",\"E05_reduced\",\n",
    "#              \"E06_F_reduced\",\"E06_M_reduced\",\"E06_reduced\"\n",
    "#              \"E07_F_reduced\",\"E07_M_reduced\",\"E07_reduced\"]\n",
    "# RESULT_SUFFIX = \"PHENO1.glm.logistic\"\n",
    "# COVARS = []\n",
    "# SPLIT_TO_TRAIN_TEST = True\n",
    "# CREATE_PHENOTYPE_FILES = False\n",
    "# CREATE_COVAR_FILE = True\n",
    "# TRAIN_RATIO = 0.8\n",
    "# input files\n",
    "PHEN_PATH = \"/cs/labs/michall/roeizucker/IIH/updated_phenotypes/ukbb_dataset_updated_11.csv\"\n",
    "GWAS_PATH = \"/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_standart_low_hwe\"\n",
    "COVAR_FILE = \"/cs/labs/michall/roeizucker/IIH/GWAS3/covariates.txt\"\n",
    "PHENS_TO_RUN = [\"H47.1\",\"G93.2\",\"both_phen\"]\n",
    "PHENOTYPES = PHENS_TO_RUN\n",
    "RESULT_SUFFIX = \"PHENO1.glm.logistic.hybrid\"\n",
    "COVARS = [\"sex\",\"year_of_birth\",\"PC1\",\"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\"]\n",
    "SPLIT_TO_TRAIN_TEST = False\n",
    "CREATE_PHENOTYPE_FILES = True\n",
    "CREATE_COVAR_FILE = True\n",
    "TRAIN_RATIO = 0.8\n",
    "NUMBER_OF_THREADS = 30\n",
    "MEMORY_AMOUNT = 24\n",
    "HADASSA_PLINK = \"hadassa\"\n",
    "REDUCED_PLINK = \"reduced\"\n",
    "ALL_PLINK = \"all\"\n",
    "HWE = 10**-50\n",
    "WHOLE_GENOME_STANDART = \"standart\"\n",
    "plink_format = REDUCED_PLINK\n",
    "\n",
    "MAF = 1e-7\n",
    "\n",
    "\n",
    "JUST_CREATE_SCRIPT = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b29cf5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_coding_low_hwe’: File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir {GWAS_PATH}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f35b59ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !head /sci/nosnap/michall/roeizucker/Alopecia_project/ukbb_dataset_updated.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6cf09519",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_files(gwas_path,phen_path,phenotypes):\n",
    "#     !mkdir {gwas_path}\n",
    "    !mkdir {gwas_path}/results\n",
    "    !mkdir {gwas_path}/phenotypes\n",
    "#     !cp /cs/labs/michall/roeizucker/10krun/runs/0:11/GWAS_delete_me/covariates.txt {gwas_path}/covariates.txt\n",
    "    !cd {gwas_path}\n",
    "    dataset = pd.read_csv(phen_path)\n",
    "    if not CREATE_PHENOTYPE_FILES:\n",
    "        return\n",
    "    for phenotype_col in phenotypes:\n",
    "        print(phenotype_col)\n",
    "        file_name = phenotype_col.lower().replace(' ', '_').replace('-', '_') + '.txt'\n",
    "        # use eid as fid, because we take one representitive of each family\n",
    "        values = dataset[['eid', 'eid', phenotype_col]].dropna()\n",
    "        if SPLIT_TO_TRAIN_TEST:\n",
    "            values_case = values[values[phenotype_col] == 1]\n",
    "            values_control = values[values[phenotype_col]== 0] \n",
    "            values_case = values_case.sample(int(len(values_case) * TRAIN_RATIO))\n",
    "            values_control = values_control.sample(int(len(values_control) * TRAIN_RATIO))\n",
    "            values = pd.concat([values_case,values_control])\n",
    "        values.to_csv(os.path.join(f\"{gwas_path}/phenotypes\", file_name), \\\n",
    "            header = False, index = False, sep = '\\t')\n",
    "    if len(COVARS) != 0:\n",
    "        covariate_values = dataset[['eid', 'eid'] + [covariate_col for covariate_col in COVARS if covariate_col != 'const']].dropna()\n",
    "        covariate_values.to_csv(os.path.join( GWAS_PATH,'covariates.txt'), header = False, index = False, sep = '\\t')\n",
    "    else:\n",
    "        ! cp {COVAR_FILE} {gwas_path}/covariates.txt\n",
    "        # An array is created with all the values needed for running GWAS.\n",
    "# The array will be sorted in batches for better file accessing\n",
    "def create_the_data_file(phenotype_cols,gwas_path,data_file_path):\n",
    "    values = []\n",
    "    last_counter = 0\n",
    "    cur_phen_counter = 0\n",
    "    counter = 0\n",
    "    for phen in phenotype_cols:\n",
    "        base_path =  gwas_path\n",
    "        for j in range(1,23):\n",
    "            if os.path.exists(os.path.join(base_path ,f\"results/{phen.lower()}_chr{j}.PHENO1.glm.linear\")) or \\\n",
    "            os.path.exists(os.path.join(base_path ,f\"results/{phen.lower()}_chr{j}.PHENO1.glm.logistic\")) or \\\n",
    "            os.path.exists(os.path.join(base_path ,f\"results/{phen.lower()}_chr{j}.{RESULT_SUFFIX}\")):\n",
    "#                 print(\"a\")\n",
    "                continue\n",
    "            \n",
    "            # TODO: change location of plink files to be const\n",
    "            if plink_format == HADASSA_PLINK:\n",
    "                values.append([counter,os.path.join(base_path, \"phenotypes/\"+ phen.lower() + \".txt\"),\n",
    "                               os.path.join(base_path, f\"results/{phen.lower()}_chr{j}\"),\n",
    "                               f\"/cs/snapless/michall/hadasak/improve_PRS/bed_files/chr{j}_imputed_snp_id_filtered\"\n",
    "                               ,os.path.join(base_path ,\"covariates.txt\"),j])\n",
    "            elif plink_format == REDUCED_PLINK:\n",
    "                values.append([counter,os.path.join(base_path, \"phenotypes/\"+ phen.lower() + \".txt\"),\n",
    "                               os.path.join(base_path, f\"results/{phen.lower()}_chr{j}\"),\n",
    "                               f\"/cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{j}\"\n",
    "                               ,os.path.join(base_path ,\"covariates.txt\"),j])\n",
    "            elif plink_format == ALL_PLINK:\n",
    "                values.append([counter,os.path.join(base_path, \"phenotypes/\"+ phen.lower() + \".txt\"),\n",
    "                               os.path.join(base_path, f\"results/{phen.lower()}_chr{j}\"),\n",
    "                               f\"/cs/biobank-genetics/EGAD00010001226/001/ukb_snp_chr{j}_v2\"\n",
    "                               ,os.path.join(base_path ,\"covariates.txt\"),j])\n",
    "            elif plink_format == WHOLE_GENOME_STANDART:\n",
    "                values.append([counter,os.path.join(base_path, \"phenotypes/\"+ phen.lower() + \".txt\"),\n",
    "                               os.path.join(base_path, f\"results/{phen.lower()}_chr{j}\"),\n",
    "                               f\"/sci/archive/michall/roeizucker/standart_genetics/chr{j}_filtered_mafe-6_hwee-5_geno_02\"\n",
    "                               ,os.path.join(base_path ,\"covariates.txt\"),j])\n",
    "                \n",
    "            counter+=1\n",
    "        cur_phen_counter+=1\n",
    "        if cur_phen_counter%BATCH_LENGTH == 0:\n",
    "            values[last_counter:counter] = sorted(values[last_counter:counter],key = lambda x:x[-1])\n",
    "            last_counter = counter\n",
    "    values[last_counter:counter] = sorted(values[last_counter:counter],key = lambda x:x[-1])\n",
    "    df = pd.DataFrame(values, columns=[\"unsorted_counter\",\"phenotype_path\",\"output_path\",\"partial_chromosome_file_path\", \"covariates\",\"chr\"])\n",
    "    df.to_csv(data_file_path)\n",
    "\n",
    "\n",
    "def write_mediator_script(mediator_script_path,data_file_path,gwas_path):\n",
    "    with open(mediator_script_path, \"w\") as mediator_file:\n",
    "        mediator_file.write(f'''import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "import os.path\n",
    "butch_num = int(sys.argv[1]) \n",
    "curr_task = int(sys.argv[2])\n",
    "BATCH_LENGTH = {BATCH_LENGTH}\n",
    "SKIP_POINT = {MAX_JOBS_BEFORE_SUBMITTING_NEW_ARRAY}\n",
    "PLINK_PATH = \"{PLINK_PATH}\"\n",
    "plink_format = \"{plink_format}\"\n",
    "data_file_path = \"{data_file_path}\"\n",
    "NUMBER_OF_THREADS = {NUMBER_OF_THREADS}\n",
    "MEMORY_AMOUNT = {MEMORY_AMOUNT}\n",
    "df = pd.read_csv(data_file_path)\n",
    "print(df.loc[butch_num * BATCH_LENGTH + curr_task])\n",
    "val = df.loc[butch_num * BATCH_LENGTH + curr_task]\n",
    "location = butch_num * BATCH_LENGTH + curr_task\n",
    "if curr_task == SKIP_POINT and location < 30500:\n",
    "    new_batch  = butch_num + 1\n",
    "    if not os.path.isfile(\"{gwas_path}_\" + str(new_batch) + \"_flag\"):\n",
    "        os.system(\"sbatch --array=0-\"+str(BATCH_LENGTH - 1)+\" --mem=\"+str(MEMORY_AMOUNT + 1)+\" -c{NUMBER_OF_THREADS}--time=3-0 --killable --requeue --wrap=\\\\\"{os.path.join( gwas_path,\"master.sh\")} \"+str(new_batch)+\"\\\\\"\")\n",
    "if location < 30500:\n",
    "    if plink_format == \"{ALL_PLINK}\":\n",
    "        print(PLINK_PATH + \" --bed \" + str(df.loc[location].partial_chromosome_file_path).replace(\"snp\",\"cal\") + \".bed --bim \" + df.loc[location].partial_chromosome_file_path + \".bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.fam --pheno \" + df.loc[location].phenotype_path + \" --covar \" + df.loc[location].covariates + \"  --out  \" + df.loc[location].output_path + \" --1 --glm hide-covar --covar-variance-standardize --freq --threads {NUMBER_OF_THREADS} --maf {MAF} --hwe {HWE} --geno 0.1 --memory \" + str(MEMORY_AMOUNT * 1000) ,flush=True)\n",
    "        os.system(PLINK_PATH + \" --bed \" + str(df.loc[location].partial_chromosome_file_path).replace(\"snp\",\"cal\") + \".bed --bim \" + df.loc[location].partial_chromosome_file_path + \".bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.fam --pheno \" + df.loc[location].phenotype_path + \" --covar \" + df.loc[location].covariates + \"  --out  \" + df.loc[location].output_path + \" --1 --glm hide-covar --covar-variance-standardize --freq --threads {NUMBER_OF_THREADS} --maf {MAF} --hwe {HWE} --geno 0.1 --memory \" +str(MEMORY_AMOUNT * 1000))\n",
    "    else:\n",
    "        print(PLINK_PATH + \" --bed \" + df.loc[location].partial_chromosome_file_path + \".bed --bim \" + df.loc[location].partial_chromosome_file_path + \".bim --fam \"+df.loc[location].partial_chromosome_file_path + \".fam --pheno \" + df.loc[location].phenotype_path + \" --covar \" + df.loc[location].covariates + \"  --out  \" + df.loc[location].output_path + \" --1 --glm hide-covar --covar-variance-standardize --freq --threads {NUMBER_OF_THREADS} --maf {MAF} --hwe {HWE} --geno 0.1 --memory \" + str(MEMORY_AMOUNT * 1000),flush=True)\n",
    "        os.system(PLINK_PATH + \" --bed \" + df.loc[location].partial_chromosome_file_path + \".bed --bim \" + df.loc[location].partial_chromosome_file_path + \".bim --fam \"+df.loc[location].partial_chromosome_file_path + \".fam --pheno \" + df.loc[location].phenotype_path + \" --covar \" + df.loc[location].covariates + \"  --out  \" + df.loc[location].output_path + \" --1 --glm hide-covar --covar-variance-standardize --freq --threads {NUMBER_OF_THREADS} --maf {MAF} --hwe {HWE} --geno 0.1 --memory \" + str(MEMORY_AMOUNT * 1000))\n",
    "    print(\"done!\")\n",
    "''')\n",
    "\n",
    "def write_master_script(master_script_path,mediator_script_path,gwas_path):\n",
    "    with open(master_script_path, \"w\") as master_file:\n",
    "        master_file.write(f'''\n",
    "FILE={gwas_path}_$1_flag\n",
    "if ! test \"$FILE\" \n",
    "then \n",
    "    touch {gwas_path}_$1_flag\n",
    "fi\n",
    "python3 {mediator_script_path} $1 $SLURM_ARRAY_TASK_ID\n",
    "''')\n",
    "    !chmod 744 {master_script_path}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d1ec7e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # data_file_path = os.path.join(GWAS_PATH, \"data_file.csv\")\n",
    "# # create_the_data_file(PHENOTYPES,GWAS_PATH,data_file_path)\n",
    "# var = \"/cs/biobank-genetics/EGAD00010001226/001/ukb_snp_chr1_v2.bed\"\n",
    "# print(var)\n",
    "# print(str(var).replace(\"snp\",\"cal\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "751ef0f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# master_script_path = os.path.join(GWAS_PATH , \"master.sh\")\n",
    "# print(f'''how to run:\n",
    "# sbatch --array=0-263 --mem=12g -c10 --time=3-0 --requeue --killable --wrap=\"{master_script_path} 0\"''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4398ce33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_coding_low_hwe/results’: File exists\n",
      "mkdir: cannot create directory ‘/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_coding_low_hwe/phenotypes’: File exists\n",
      "H47.1\n",
      "G93.2\n",
      "both_phen\n",
      "how to run:\n",
      "sbatch --array=0-263 --mem=25g -c30 --time=5-0 --requeue --killable --wrap=\"/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_coding_low_hwe/master.sh 0\"\n"
     ]
    }
   ],
   "source": [
    "# script body (will be transferred to another file)\n",
    "\n",
    "\n",
    "# TODO: add verbous argument, and better explain the output\n",
    "\n",
    "# set input \n",
    "# TODO: change so it is accepted as params\n",
    "phen_path = PHEN_PATH\n",
    "gwas_path = GWAS_PATH\n",
    "\n",
    "# do stuff\n",
    "phen_dataset = pd.read_csv(PHEN_PATH)\n",
    "phenotype_cols = list(phen_dataset)\n",
    "phenotype_cols = phenotype_cols[2:phenotype_cols.index(VALUE_AT_END_OF_PHENOTYPES)]\n",
    "\n",
    "# we will use the phenotypes parameter only if it exists\n",
    "if len (PHENOTYPES) > 0:\n",
    "    phenotype_cols = PHENOTYPES\n",
    "    \n",
    "# TODO: chnage so names are consts\n",
    "data_file_path = os.path.join(gwas_path, \"data_file.csv\")\n",
    "mediator_script_path = os.path.join(gwas_path , \"mediator.py\")\n",
    "master_script_path = os.path.join(gwas_path , \"master.sh\")\n",
    "\n",
    "create_files(gwas_path,phen_path,phenotype_cols)\n",
    "create_the_data_file(phenotype_cols,gwas_path,data_file_path)\n",
    "write_mediator_script(mediator_script_path,data_file_path,gwas_path)\n",
    "write_master_script(master_script_path,mediator_script_path,gwas_path)\n",
    "\n",
    "# TODO: add output folder\n",
    "print(f'''how to run:\n",
    "sbatch --array=0-263 --mem={MEMORY_AMOUNT + 1}g -c{NUMBER_OF_THREADS} --time=5-0 --requeue --killable --wrap=\"{master_script_path} 0\"''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2133a296-c034-49bd-bcbf-2e6e90189609",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H47.1\n",
      "both 91\n",
      "female 53\n",
      "male 38\n",
      "G93.2\n",
      "both 116\n",
      "female 77\n",
      "male 39\n",
      "both_phen\n",
      "both 189\n",
      "female 115\n",
      "male 74\n",
      "covars (10) ['sex', 'year_of_birth', 'PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "    \n",
    "# print(phen_dataset[PHENS_TO_RUN].describe())\n",
    "for phen in PHENS_TO_RUN:\n",
    "    print(phen)\n",
    "    print(\"both\",len(phen_dataset[phen_dataset[phen] == 1]))\n",
    "    print(\"female\",len(phen_dataset[(phen_dataset[phen] == 1) &(phen_dataset[\"sex\"] == 0)  ]))\n",
    "    print(\"male\",len(phen_dataset[(phen_dataset[phen] == 1) &(phen_dataset[\"sex\"] == 1)]))\n",
    "\n",
    "print(F\"covars ({len(COVARS)})\",COVARS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c9fc88b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_the_data_file(phenotype_cols,gwas_path,data_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d37fe7bc",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid decimal literal (3184066541.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[12], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    sbatch --array=0-3 --mem=25g -c30 --time=5-0 --requeue --killable --wrap=\"/sci/archive/michall/roeizucker/IIH/GWAS_standart/master.sh 0\"\u001b[0m\n\u001b[0m                              ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid decimal literal\n"
     ]
    }
   ],
   "source": [
    "\n",
    "sbatch --array=0-3 --mem=25g -c30 --time=5-0 --requeue --killable --wrap=\"/sci/archive/michall/roeizucker/IIH/GWAS_standart/master.sh 0\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53754189",
   "metadata": {},
   "outputs": [],
   "source": [
    "# master_script_path = priority_gwas_dir + \"master.sh\"\n",
    "\n",
    "# (mediator_script_path,data_file_path,gwas_path)\n",
    "master_script_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54fb650a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv(data_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0810d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'''how to run:\n",
    "sbatch --array=0-263 --mem=12g -c10 --time=3-0ס --wrap=\"{master_script_path} 0\"''')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f742d56a",
   "metadata": {},
   "source": [
    "## combines GWAS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b554bf38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_standart_low_hwe/results/H47.1.csv\n",
      "/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_standart_low_hwe/results/G93.2.csv\n",
      "/cs/labs/michall/roeizucker/IIH/updated_phenotypes/GWAS_standart_low_hwe/results/both_phen.csv\n"
     ]
    }
   ],
   "source": [
    "import glob, os\n",
    "# path = \"/cs/labs/michall/roeizucker/10krun/runs/48:59/GWAS/results\"\n",
    "# os.chdir(path)\n",
    "path = os.path.join(GWAS_PATH,\"results\")\n",
    "# files = []\n",
    "# for file in glob.glob(\"*.glm.*\"):\n",
    "#     files.append(file)\n",
    "phens = PHENOTYPES\n",
    "for phen in phens:\n",
    "    dataframes = []\n",
    "    for j in range(1,23):\n",
    "#         change so logistic/liner is detrmined generically\n",
    "        file_name = os.path.join(path,f\"{phen.lower()}_chr{j}.{RESULT_SUFFIX}\")\n",
    "        try:\n",
    "            df = pd.read_csv(file_name,sep=\"\\t\")\n",
    "            dataframes.append(df)\n",
    "        except pd.errors.EmptyDataError:\n",
    "            print(file_name)\n",
    "            continue\n",
    "        except FileNotFoundError:\n",
    "            print(file_name)\n",
    "            continue\n",
    "    if len(dataframes) == 0:\n",
    "        continue\n",
    "    master_df = pd.concat(dataframes)\n",
    "    master_df.to_csv(f\"{os.path.join(path,phen)}.csv\",index=False)\n",
    "    print(f\"{os.path.join(path,phen)}.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6661d74f",
   "metadata": {},
   "source": [
    "### combine single phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e966a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = os.path.join(GWAS_PATH,\"results\")\n",
    "phen = \"L63\"\n",
    "dataframes = []\n",
    "for j in range(1,23):\n",
    "#         change so logistic/liner is detrmined generically\n",
    "    file_name = os.path.join(path,f\"{phen.lower()}_chr{j}.{RESULT_SUFFIX}\")\n",
    "    try:\n",
    "        df = pd.read_csv(file_name,sep=\"\\t\")\n",
    "        dataframes.append(df)\n",
    "    except pd.errors.EmptyDataError:\n",
    "        print(file_name)\n",
    "    except FileNotFoundError:\n",
    "        print(file_name)\n",
    "if not len(dataframes) == 0:\n",
    "    master_df = pd.concat(dataframes)\n",
    "    master_df.to_csv(f\"{os.path.join(path,phen)}.csv\",index=False)\n",
    "    print(f\"{os.path.join(path,phen)}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f3451d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.path.join(path,phen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da943c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#delete this\n",
    "df = pd.read_csv(\"/sci/archive/michall/roeizucker/IIH/GWAS_standart/results/updated_PAP_genes1.csv\")\n",
    "df = df[df[\"P\"] < 5*10**-5]\n",
    "df.to_csv(\"/sci/archive/michall/roeizucker/IIH/GWAS_standart/results/updated_PAP_with_genes_e-5.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77086d5f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/sci/archive/michall/roeizucker/IIH/GWAS_standart/results/updated_Both_with_genes_e-5.csv\")\n",
    "df[\"gene_symbol\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "044465d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "/sci/archive/michall/roeizucker/IIH/GWAS_standart/results/*_genes_e-5.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98c425bb",
   "metadata": {},
   "source": [
    "# Run GWAS for specific failed cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39980688",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# gwas_path = GWAS_PATH\n",
    "\n",
    "os.path.join(GWAS_PATH,\"data_file.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942c4b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "FAILED_FILES =[\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/D12_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/F171_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/I209_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/I20_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/I25_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/I48_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K219_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K52_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K52_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K52_chr3.PHENO1.glm.logistic\", \n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K635_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K63_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K63_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K63_chr3.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K80_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K80_chr15.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/K80_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/M139_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/M13_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/M13_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/N40_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_155.1_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/220:229/GWAS/results/phecode_155.1_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_159.4_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/240:249/GWAS/results/phecode_159.4_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/260:269/GWAS/results/phecode_172.2_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/260:269/GWAS/results/phecode_172.2_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/260:269/GWAS/results/phecode_172.2_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/260:269/GWAS/results/phecode_172.2_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_172.2_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_172.2_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_172.2_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_172.2_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_198.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/370:379/GWAS/results/phecode_198.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/410:419/GWAS/results/phecode_202.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_202.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/420:429/GWAS/results/phecode_204.0_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_204.0_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/450:459/GWAS/results/phecode_208.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_208.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/450:459/GWAS/results/phecode_208.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_208.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_210.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/450:459/GWAS/results/phecode_210.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_211.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/460:469/GWAS/results/phecode_211.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_214.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/470:479/GWAS/results/phecode_214.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_214.0_chr5.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/470:479/GWAS/results/phecode_214.0_chr5.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_215.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/470:479/GWAS/results/phecode_215.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/480:489/GWAS/results/phecode_216.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_216.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/620:629/GWAS/results/phecode_250.2_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/620:629/GWAS/results/phecode_250.2_chr5.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/530:539/GWAS/results/phecode_228.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_262.0_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_250.2_chr5.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_255.21_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_276.14_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_250.2_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_261.2_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_276.11_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_260.6_chr10.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_260.6_chr10.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_275.5_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_261.2_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_262.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_276.14_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_276.14_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_272.1_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_228.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/700:709/GWAS/results/phecode_255.21_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/780:789/GWAS/results/phecode_262.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/780:789/GWAS/results/phecode_262.0_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/870:879/GWAS/results/phecode_275.5_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/830:839/GWAS/results/phecode_272.1_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/770:779/GWAS/results/phecode_261.2_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/770:779/GWAS/results/phecode_261.2_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/880:889/GWAS/results/phecode_276.11_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/760:769/GWAS/results/phecode_260.6_chr10.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/760:769/GWAS/results/phecode_260.6_chr10.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/890:899/GWAS/results/phecode_276.14_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/890:899/GWAS/results/phecode_276.14_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/890:899/GWAS/results/phecode_276.14_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_287.3_chr12.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_292.3_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_292.3_chr12.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_290.1_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_292.3_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_290.1_chr10.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_290.1_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1150:1159/GWAS/results/phecode_292.3_chr12.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1150:1159/GWAS/results/phecode_292.3_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1150:1159/GWAS/results/phecode_292.3_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1110:1119/GWAS/results/phecode_290.1_chr10.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1110:1119/GWAS/results/phecode_290.1_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1110:1119/GWAS/results/phecode_290.1_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1060:1069/GWAS/results/phecode_287.3_chr12.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1180:1189/GWAS/results/phecode_296.1_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1180:1189/GWAS/results/phecode_296.1_chr13.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_296.1_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_296.1_chr13.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1420:1429/GWAS/results/phecode_344.0_chr13.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1420:1429/GWAS/results/phecode_344.0_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/1420:1429/GWAS/results/phecode_344.0_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2260:2269/GWAS/results/phecode_458.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2260:2269/GWAS/results/phecode_458.0_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2260:2269/GWAS/results/phecode_458.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2340:2349/GWAS/results/phecode_478.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/30:39/GWAS/results/phecode_41.1_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2600:2609/GWAS/results/phecode_528.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2600:2609/GWAS/results/phecode_528.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_458.0_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_344.0_chr13.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_458.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_497.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_458.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_441.0_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_441.0_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_41.1_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_433.8_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_433.8_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_528.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_528.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_478.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_344.0_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_427.12_chr3.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_344.0_chr7.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_427.41_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_427.41_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2140:2149/GWAS/results/phecode_441.0_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2140:2149/GWAS/results/phecode_441.0_chr8.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2390:2399/GWAS/results/phecode_497.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2040:2049/GWAS/results/phecode_427.12_chr3.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2130:2139/GWAS/results/phecode_433.8_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2130:2139/GWAS/results/phecode_433.8_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2050:2059/GWAS/results/phecode_427.41_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2050:2059/GWAS/results/phecode_427.41_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2920:2929/GWAS/results/phecode_578.1_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2920:2929/GWAS/results/phecode_578.1_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2920:2929/GWAS/results/phecode_578.1_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2990:2999/GWAS/results/phecode_587.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3030:3039/GWAS/results/phecode_593.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3030:3039/GWAS/results/phecode_593.0_chr3.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3030:3039/GWAS/results/phecode_593.0_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_559.0_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_578.1_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_559.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_593.0_chr3.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_556.1_chr16.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_593.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_530.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_530.7_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_530.7_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_578.1_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_578.1_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_593.0_chr14.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_587.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_613.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_618.1_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2760:2769/GWAS/results/phecode_559.0_chr9.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2760:2769/GWAS/results/phecode_559.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2660:2669/GWAS/results/phecode_530.7_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2660:2669/GWAS/results/phecode_530.7_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2630:2639/GWAS/results/phecode_530.0_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3240:3249/GWAS/results/phecode_618.1_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3180:3189/GWAS/results/phecode_613.0_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/2750:2759/GWAS/results/phecode_556.1_chr16.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3840:3849/GWAS/results/phecode_716.9_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3520:3529/GWAS/results/phecode_671.0_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3570:3579/GWAS/results/phecode_687.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3270:3279/GWAS/results/phecode_619.4_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3270:3279/GWAS/results/phecode_619.4_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3270:3279/GWAS/results/phecode_619.4_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3270:3279/GWAS/results/phecode_619.4_chr16.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3270:3279/GWAS/results/phecode_619.4_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3270:3279/GWAS/results/phecode_619.4_chr15.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3990:3999/GWAS/results/phecode_735.21_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_703.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_716.9_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_741.2_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_626.12_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_681.2_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_626.12_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_619.4_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_619.4_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_619.4_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_619.4_chr15.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_671.0_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_687.0_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_735.21_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_619.4_chr16.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_626.12_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_619.4_chr6.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4050:4059/GWAS/results/phecode_741.2_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3700:3709/GWAS/results/phecode_703.0_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3540:3549/GWAS/results/phecode_681.2_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3320:3329/GWAS/results/phecode_626.12_chr11.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3320:3329/GWAS/results/phecode_626.12_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3320:3329/GWAS/results/phecode_626.12_chr19.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4160:4169/GWAS/results/phecode_751.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4550:4559/GWAS/results/phecode_960.0_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_747.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_742.8_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_751.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_960.0_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_743.2_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_747.1_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4070:4079/GWAS/results/phecode_742.8_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4100:4109/GWAS/results/phecode_747.11_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4100:4109/GWAS/results/phecode_747.1_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4080:4089/GWAS/results/phecode_743.2_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/R69_chr12.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_276.14_f_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/890:899/GWAS/results/phecode_276.14_f_chr17.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3520:3529/GWAS/results/phecode_671.0_f_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_671.0_f_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_741.2_f_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_681.2_m_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_741.2_m_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4050:4059/GWAS/results/phecode_741.2_f_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4050:4059/GWAS/results/phecode_741.2_m_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/3540:3549/GWAS/results/phecode_681.2_m_chr2.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_747.11_f_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_747.11_m_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/GWAS_results/phecode_742.8_m_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4070:4079/GWAS/results/phecode_742.8_m_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4100:4109/GWAS/results/phecode_747.11_f_chr1.PHENO1.glm.logistic\",\n",
    "\"/cs/labs/michall/roeizucker/phewas_run/runs1/4100:4109/GWAS/results/phecode_747.11_m_chr1.PHENO1.glm.logistic\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11795e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "desc_file = \"/cs/labs/michall/roeizucker/phewas_run/description12.csv\"\n",
    "df = pd.read_csv(desc_file)\n",
    "df[\"name\"] = df[\"name\"]\n",
    "phen_dic = df.set_index(\"name\").to_dict()[\"dataset_path\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5319c3b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "phen_dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68d895ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "values = []\n",
    "for i,file in enumerate(FAILED_FILES):\n",
    "    chrom = re.findall(\"chr\\d+\",file)[0].replace(\"chr\",\"\")\n",
    "    phen_temp = re.findall(\"results/.+chr\",file)\n",
    "    if len(phen_temp) == 0:\n",
    "        print(\"AAA\",file)\n",
    "        continue\n",
    "    phenotype = phen_temp[0].replace(\"_chr\",\"\").replace(\"results/\",\"\")\n",
    "#     print(chrom,phenotype)\n",
    "    phen_key = phenotype\n",
    "    if phenotype[0] == \"p\":\n",
    "        phen_key = \"P\" + phen_key[1:].replace(\"_m\",\"_M\").replace(\"_f\",\"_F\")\n",
    "    gwas_path = phen_dic[phen_key].replace(\"ukbb_dataset.csv\",\"\") + \"GWAS\"\n",
    "    base_path = gwas_path\n",
    "    values.append([i,os.path.join(base_path, \"phenotypes/\"+ phenotype.lower() + \".txt\"),\n",
    "               os.path.join(base_path, f\"results/{phenotype.lower()}_chr{chrom}\"),\n",
    "               f\"/cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{chrom}\"\n",
    "               ,os.path.join(base_path ,\"covariates.txt\"),chrom])\n",
    "\n",
    "#     break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1765e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_file_path = \"/sci/nosnap/michall/roeizucker/emergancy_PWAS_HUB/data_file.csv\"\n",
    "df_data = pd.DataFrame(values, columns=[\"unsorted_counter\",\"phenotype_path\",\"output_path\",\"partial_chromosome_file_path\", \"covariates\",\"chr\"])\n",
    "df_data\n",
    "df_data.to_csv(data_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d5efd3a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9675558",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_path = \"/sci/nosnap/michall/roeizucker/emergancy_PWAS_HUB/\"\n",
    "\n",
    "mediator_script_path = os.path.join(gwas_path , \"mediator.py\")\n",
    "master_script_path = os.path.join(gwas_path , \"master.sh\")\n",
    "write_mediator_script(mediator_script_path,data_file_path,gwas_path)\n",
    "write_master_script(master_script_path,mediator_script_path,gwas_path)\n",
    "\n",
    "# TODO: add output folder\n",
    "print(f'''how to run:\n",
    "sbatch --array=0-263 --mem={MEMORY_AMOUNT + 1}g -c{NUMBER_OF_THREADS} --time=5-0 --requeue --killable --wrap=\"{master_script_path} 0\"''')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
