{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SKAT Enviroment Preperation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Descpreription\n",
    "this notebook prepares the neccesary files and scripts to run the SKAT analysis for a given phenotype (in csv file) and plink format genetic files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pandas_plink as pdp\n",
    "from ipywidgets import IntProgress\n",
    "from IPython import display\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_PATH = \"/cs/labs/michall/roeizucker/SKAT_experiments/SetID\"\n",
    "FILE_CONTAINING_PARTICIPANTS = PROJECT_PATH + \"/data_participant.csv\"\n",
    "OLD_PHENOTYPE_FILE = '/cs/labs/michall/roeizucker/10krun/runs/48:59/ukbb_dataset.csv'\n",
    "PHENOTYPE_FILE_NAME = PROJECT_PATH + \"/phens.csv\"\n",
    "SET_ID_FILE_MAX_LENGTH = 5000\n",
    "GENE_FILE_PATH = '/cs/labs/michall/roeizucker/virt_env_install_test/genes_hg19.csv'\n",
    "# temp covar /cs/labs/michall/roeizucker/phewas_run/runs1/1000:1009/GWAS/covariates.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create basic SetID file \n",
    "the SKAT program uses a SetID file that defines which gene contains which variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENES_FILE_PATH = \"/cs/usr/roeizucker/my_storage/virt_env_install_test/genes_hg19.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load genes file\n",
    "this file is part of the PWAS pipeline, it defines the location and size of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_file_dataframe = pd.read_csv(GENES_FILE_PATH)[[\"uniprot_id\",\"chr\",\"cds_start\",\"cds_end\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load PLINK format genetic files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "PLINK_FILE_BASE_PATH = \"/cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_results = []\n",
    "for i in range(1,23):\n",
    "    df = pdp.read_plink(PLINK_FILE_BASE_PATH + str(i),verbose=False)[0][[\"snp\",\"pos\"]]\n",
    "    result = [(x, y) for x, y in zip(df['snp'], df['pos'])]\n",
    "    plink_results.append((i,result))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n"
     ]
    }
   ],
   "source": [
    "print(len(plink_results))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr is: 22\n"
     ]
    }
   ],
   "source": [
    "print(\"chr is:\",i)\n",
    "val = pdp.read_plink(PLINK_FILE_BASE_PATH + str(i),verbose=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<tr>\n",
       "<td>\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr><td> </td><th> Array </th><th> Chunk </th></tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr><th> Bytes </th><td> 27.24 GiB </td> <td> 4.00 MiB </td></tr>\n",
       "    <tr><th> Shape </th><td> (15002, 487409) </td> <td> (1024, 1024) </td></tr>\n",
       "    <tr><th> Count </th><td> 42840 Tasks </td><td> 7140 Chunks </td></tr>\n",
       "    <tr><th> Type </th><td> float32 </td><td> numpy.ndarray </td></tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</td>\n",
       "<td>\n",
       "<svg width=\"170\" height=\"81\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"2\" x2=\"120\" y2=\"2\" />\n",
       "  <line x1=\"0\" y1=\"4\" x2=\"120\" y2=\"4\" />\n",
       "  <line x1=\"0\" y1=\"6\" x2=\"120\" y2=\"6\" />\n",
       "  <line x1=\"0\" y1=\"8\" x2=\"120\" y2=\"8\" />\n",
       "  <line x1=\"0\" y1=\"10\" x2=\"120\" y2=\"10\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"120\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"15\" x2=\"120\" y2=\"15\" />\n",
       "  <line x1=\"0\" y1=\"17\" x2=\"120\" y2=\"17\" />\n",
       "  <line x1=\"0\" y1=\"19\" x2=\"120\" y2=\"19\" />\n",
       "  <line x1=\"0\" y1=\"21\" x2=\"120\" y2=\"21\" />\n",
       "  <line x1=\"0\" y1=\"23\" x2=\"120\" y2=\"23\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" />\n",
       "  <line x1=\"0\" y1=\"27\" x2=\"120\" y2=\"27\" />\n",
       "  <line x1=\"0\" y1=\"30\" x2=\"120\" y2=\"30\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"120\" y2=\"31\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"31\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"6\" y1=\"0\" x2=\"6\" y2=\"31\" />\n",
       "  <line x1=\"12\" y1=\"0\" x2=\"12\" y2=\"31\" />\n",
       "  <line x1=\"18\" y1=\"0\" x2=\"18\" y2=\"31\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"31\" y1=\"0\" x2=\"31\" y2=\"31\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"31\" />\n",
       "  <line x1=\"44\" y1=\"0\" x2=\"44\" y2=\"31\" />\n",
       "  <line x1=\"50\" y1=\"0\" x2=\"50\" y2=\"31\" />\n",
       "  <line x1=\"56\" y1=\"0\" x2=\"56\" y2=\"31\" />\n",
       "  <line x1=\"63\" y1=\"0\" x2=\"63\" y2=\"31\" />\n",
       "  <line x1=\"69\" y1=\"0\" x2=\"69\" y2=\"31\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"31\" />\n",
       "  <line x1=\"81\" y1=\"0\" x2=\"81\" y2=\"31\" />\n",
       "  <line x1=\"88\" y1=\"0\" x2=\"88\" y2=\"31\" />\n",
       "  <line x1=\"94\" y1=\"0\" x2=\"94\" y2=\"31\" />\n",
       "  <line x1=\"100\" y1=\"0\" x2=\"100\" y2=\"31\" />\n",
       "  <line x1=\"107\" y1=\"0\" x2=\"107\" y2=\"31\" />\n",
       "  <line x1=\"113\" y1=\"0\" x2=\"113\" y2=\"31\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"31\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,31.544031670163466 0.0,31.544031670163466\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"51.544032\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >487409</text>\n",
       "  <text x=\"140.000000\" y=\"15.772016\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,140.000000,15.772016)\">15002</text>\n",
       "</svg>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "dask.array<transpose, shape=(15002, 487409), dtype=float32, chunksize=(1024, 1024), chunktype=numpy.ndarray>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = val[2]\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create gene-variant associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n",
      "99.98666844420744\n"
     ]
    }
   ],
   "source": [
    "gene_association_result = []\n",
    "\n",
    "for (chrom, vals) in plink_results:\n",
    "    reduced_gene_dataframe = gene_file_dataframe[gene_file_dataframe[\"chr\"] == str(chrom)]\n",
    "    display.clear_output()\n",
    "    print(chrom)\n",
    "    counter = 0\n",
    "    for val in vals:\n",
    "        pos = val[1]\n",
    "        name= val[0]\n",
    "        genes = list(reduced_gene_dataframe[(reduced_gene_dataframe[\"cds_start\"] <= pos) & (reduced_gene_dataframe[\"cds_end\"] >= pos)][\"uniprot_id\"])\n",
    "        for gene in genes:\n",
    "            gene_association_result.append((name,gene))\n",
    "        counter +=1\n",
    "        if counter %3000 == 0:\n",
    "            display.clear_output()\n",
    "            print(chrom)\n",
    "            print(counter/len(vals) * 100)\n",
    "gene_association_result = sorted(gene_association_result,key= lambda x:x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "651400\n"
     ]
    }
   ],
   "source": [
    "print(len(gene_association_result))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write variant-gene associations to SetID file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49.12496162112374\n"
     ]
    }
   ],
   "source": [
    "set_id_file_path = PROJECT_PATH + \"/\" + \"genes.SetID\"\n",
    "counter = 0\n",
    "lines_seen = set()\n",
    "counter = 0\n",
    "with open(set_id_file_path,\"w\") as file:\n",
    "    for (variant,gene) in gene_association_result[0:len(gene_association_result) // 2]:\n",
    "        if counter %10000 == 0:\n",
    "            display.clear_output()\n",
    "            print(counter/len(gene_association_result) * 100)\n",
    "        if f\"{gene} {variant}\" not in lines_seen:\n",
    "            file.write(f\"{gene} {variant}\\n\")\n",
    "            lines_seen.add(f\"{gene} {variant}\")  \n",
    "        counter +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/cs/labs/michall/roeizucker/SKAT_experiments/SetID'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PROJECT_PATH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write variant-gene associations to multiple SetID files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99.78507829290758\n"
     ]
    }
   ],
   "source": [
    "def get_variants_for_genes(counter,gene_association_result):\n",
    "    new_counter = counter\n",
    "    curr_gene = gene_association_result[new_counter][1]    \n",
    "    gene_vars = []\n",
    "    while new_counter < len(gene_association_result) and gene_association_result[new_counter][1] == curr_gene:\n",
    "        gene_vars.append(gene_association_result[new_counter])\n",
    "        new_counter+=1\n",
    "        if new_counter %10000 == 0:\n",
    "            display.clear_output()\n",
    "            print(new_counter/len(gene_association_result) * 100)\n",
    "    return gene_vars,new_counter\n",
    "\n",
    "\n",
    "set_id_file_path = PROJECT_PATH + \"/\" + \"genes.SetID\"\n",
    "counter = 0\n",
    "lines_seen = set()\n",
    "counter = 0\n",
    "curr_file = 0\n",
    "add_to_curr_file = []\n",
    "while counter < len(gene_association_result):\n",
    "    curr_gene = gene_association_result[counter][1]\n",
    "    add_to_curr_file,counter = get_variants_for_genes(counter,gene_association_result)\n",
    "    new_counter = counter\n",
    "    gene_vars = []\n",
    "    while len(add_to_curr_file) + len(gene_vars) < SET_ID_FILE_MAX_LENGTH and new_counter < len(gene_association_result):\n",
    "        \n",
    "        add_to_curr_file.extend(gene_vars)\n",
    "        counter = new_counter\n",
    "        if new_counter >= len(gene_association_result):\n",
    "            print(\"aaaa\")\n",
    "        gene_vars,new_counter = get_variants_for_genes(counter,gene_association_result)\n",
    "        \n",
    "#     set_id_file_path = PROJECT_PATH + \"/\" + \"genes.SetID\"\n",
    "    set_id_file_path = f\"{PROJECT_PATH}/genes_{curr_file}.SetID\"\n",
    "    with open(set_id_file_path,\"w\") as file:\n",
    "        for (variant,gene) in add_to_curr_file:\n",
    "            if f\"{gene} {variant}\" not in lines_seen:\n",
    "                file.write(f\"{gene} {variant}\\n\")\n",
    "                lines_seen.add(f\"{gene} {variant}\")  \n",
    "    \n",
    "    curr_file+=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create slurm scripts for generating SSD from setID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only for a single chromosome:\n",
    "CHROM = 4\n",
    "SET_ID_PATH = \"/cs/labs/michall/roeizucker/SKAT_experiments/SetID\"\n",
    "R_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{CHROM}.R\"\n",
    "PYTHON_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{CHROM}.py\"\n",
    "CSV_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{CHROM}.csv\"\n",
    "BASH_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{CHROM}.sh\"\n",
    "BASE_PATH = \"/cs/labs/michall/roeizucker/IIH\"\n",
    "CURR_BED = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{CHROM}.bed\"\n",
    "CURR_BIM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{CHROM}.bim\"\n",
    "CURR_FAM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{CHROM}.fam\"\n",
    "with open(R_FILE,\"w\") as file:\n",
    "    file.write('''require(SKAT)\n",
    "args = commandArgs(trailingOnly=TRUE)\n",
    "File.Bed<-args[1]\n",
    "File.Bim<-args[2]\n",
    "File.Fam<-args[3]\n",
    "File.SetID<-args[4]\n",
    "File.SSD<-args[5]\n",
    "File.Info<-args[6]\n",
    "Generate_SSD_SetID(File.Bed, File.Bim, File.Fam, File.SetID, File.SSD, File.Info)\n",
    "print(\"Done\")\n",
    "''')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(CSV_FILE,\"w\") as file:\n",
    "    file.write(\"bed,bim,fam,SetID,SSD,info\\n\")\n",
    "    for i in range(curr_file):\n",
    "        file.write(f\"{CURR_BED},{CURR_BIM},{CURR_FAM},{SET_ID_PATH}/genes_{i}.SetID,{BASE_PATH}/genes_{CHROM}_{i}.SSD,{BASE_PATH}/genes_{CHROM}_{i}.SSD.info\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '-f'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-205-c9b690287048>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCSV_FILE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;31m# line = df.iloc[3,]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/cs/cbio/netanel/lib/R4/bin/Rscript '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mR_FILE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbed\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbim\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfam\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSetID\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSD\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: '-f'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "import os\n",
    "df = pd.read_csv(CSV_FILE)\n",
    "line = df.iloc[int(sys.argv[1]),]\n",
    "# line = df.iloc[3,]\n",
    "os.system('/cs/cbio/netanel/lib/R4/bin/Rscript ' + R_FILE + ' ' + line.bed + ' ' + line.bim + ' ' + line.fam + ' ' + line.SetID + ' ' + line.SSD + ' ' + line.info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(PYTHON_FILE,\"w\") as file:\n",
    "    file.write(\"import pandas as pd\\n\")\n",
    "    file.write(\"import sys\\n\")\n",
    "    file.write(\"import os\\n\")\n",
    "    file.write(\"df = pd.read_csv('\"+CSV_FILE+\"')\\n\")\n",
    "    file.write(\"line = df.iloc[int(sys.argv[1]),]\\n\")\n",
    "    file.write(\"os.system('/cs/cbio/netanel/lib/R4/bin/Rscript \" + R_FILE + \" ' + line.bed + ' ' + line.bim + ' ' + line.fam + ' ' + line.SetID + ' ' + line.SSD + ' ' + line.info)\\n\")\n",
    "    file.write(\"\\n\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(BASH_FILE,\"w\") as file:\n",
    "    file.write(f\"python {PYTHON_FILE} $SLURM_ARRAY_TASK_ID\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "!chmod 744 {BASH_FILE}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_4.sh\r\n"
     ]
    }
   ],
   "source": [
    "!echo {BASH_FILE}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create SETId files for multiple chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_0.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_1.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_2.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_3.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_4.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_5.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_6.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_7.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_8.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_9.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_10.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_11.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_12.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_13.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_14.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_15.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_16.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_17.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_18.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_19.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_20.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_21.sh\n",
      "/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_22.sh\n"
     ]
    }
   ],
   "source": [
    "for chrom in range(0,23):\n",
    "    SET_ID_PATH = \"/cs/labs/michall/roeizucker/SKAT_experiments/SetID\"\n",
    "    R_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{chrom}.R\"\n",
    "    PYTHON_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{chrom}.py\"\n",
    "    CSV_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{chrom}.csv\"\n",
    "    BASH_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{chrom}.sh\"\n",
    "    BASE_PATH = \"/cs/labs/michall/roeizucker/IIH\"\n",
    "    CURR_BED = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{chrom}.bed\"\n",
    "    CURR_BIM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{chrom}.bim\"\n",
    "    CURR_FAM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{chrom}.fam\"\n",
    "    with open(R_FILE,\"w\") as file:\n",
    "        file.write('''require(SKAT)\n",
    "    args = commandArgs(trailingOnly=TRUE)\n",
    "    File.Bed<-args[1]\n",
    "    File.Bim<-args[2]\n",
    "    File.Fam<-args[3]\n",
    "    File.SetID<-args[4]\n",
    "    File.SSD<-args[5]\n",
    "    File.Info<-args[6]\n",
    "    Generate_SSD_SetID(File.Bed, File.Bim, File.Fam, File.SetID, File.SSD, File.Info)\n",
    "    print(\"abcd\")\n",
    "    ''')\n",
    "    with open(CSV_FILE,\"w\") as file:\n",
    "        file.write(\"bed,bim,fam,SetID,SSD,info\\n\")\n",
    "        for i in range(curr_file):\n",
    "            file.write(f\"{CURR_BED},{CURR_BIM},{CURR_FAM},{SET_ID_PATH}/genes_{i}.SetID,{BASE_PATH}/genes_{chrom}_{i}.SSD,{BASE_PATH}/genes_{chrom}_{i}.SSD.info\\n\")\n",
    "    with open(PYTHON_FILE,\"w\") as file:\n",
    "        file.write(\"import pandas as pd\\n\")\n",
    "        file.write(\"import sys\\n\")\n",
    "        file.write(\"import os\\n\")\n",
    "        file.write(\"df = pd.read_csv('\"+CSV_FILE+\"')\\n\")\n",
    "        file.write(\"line = df.iloc[int(sys.argv[1]),]\\n\")\n",
    "        file.write(\"os.system('/cs/cbio/netanel/lib/R4/bin/Rscript \" + R_FILE + \" ' + line.bed + ' ' + line.bim + ' ' + line.fam + ' ' + line.SetID + ' ' + line.SSD + ' ' + line.info)\\n\")\n",
    "        file.write(\"\\n\")\n",
    "    \n",
    "    with open(BASH_FILE,\"w\") as file:\n",
    "        file.write(f\"python {PYTHON_FILE} $SLURM_ARRAY_TASK_ID\\n\")\n",
    "\n",
    "    !chmod 744 {BASH_FILE}\n",
    "    !echo {BASH_FILE}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_0.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/IIH/scripts/create_SSD_0.sh running\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_1.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/IIH/scripts/create_SSD_1.sh running\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_2.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/IIH/scripts/create_SSD_2.sh running\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_3.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/IIH/scripts/create_SSD_3.sh running\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_4.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/IIH/scripts/create_SSD_4.sh running\n"
     ]
    }
   ],
   "source": [
    "for chrom in range(0,5):\n",
    "    BASH_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{chrom}.sh\"\n",
    "    print(f'sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"{BASH_FILE}\"')\n",
    "    print(\"sleep 1m\")\n",
    "    print(f\"echo {BASH_FILE} running\")\n",
    "#     print(\"sleep 24m\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create output files from SSD for all chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_0.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_1.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_2.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_3.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_4.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_5.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_6.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_7.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_8.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_9.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_10.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_11.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_12.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_13.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_14.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_15.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_16.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_17.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_18.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_19.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_20.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_21.sh\n",
      "/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_22.sh\n"
     ]
    }
   ],
   "source": [
    "for chrom in range(1,23):\n",
    "    SET_ID_PATH = \"/cs/labs/michall/roeizucker/IIH\"\n",
    "    CSV_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{chrom}.csv\"\n",
    "    BASE_PATH = \"/cs/labs/michall/roeizucker/IIH\"\n",
    "    CURR_BED = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{chrom}.bed\"\n",
    "    CURR_BIM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{chrom}.bim\"\n",
    "    CURR_FAM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{chrom}.fam\"\n",
    "    \n",
    "    R_FILE = f\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_{chrom}.R\"\n",
    "    PYTHON_FILE = f\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_{chrom}.py\"\n",
    "    # CSV_FILE = \"/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/scripts/create_SSD.csv\"\n",
    "    BASH_FILE = f\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_{chrom}.sh\"\n",
    "    with open(R_FILE,\"w\") as file:\n",
    "        file.write('''require(SKAT)\n",
    "args = commandArgs(trailingOnly=TRUE)\n",
    "File.Bed<-args[1]\n",
    "File.Bim<-args[2]\n",
    "File.Fam<-args[3]\n",
    "File.SetID<-args[4]\n",
    "File.SSD<-args[5]\n",
    "File.Info<-args[6]\n",
    "File.Cov<-\"/cs/labs/michall/roeizucker/phewas_run/runs1/1000:1009/GWAS/covariates.txt\"\n",
    "\n",
    "FAM_Cov<-Read_Plink_FAM_Cov(File.Fam, File.Cov, Is.binary=FALSE, cov_header=FALSE)\n",
    "\n",
    "X2 = FAM_Cov$Sex\n",
    "X1 = FAM_Cov$COV1\n",
    "\n",
    "SSD.INFO<-Open_SSD(File.SSD, File.Info)\n",
    "\n",
    "test_phenotypes = read.csv(file = \"/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/phens.csv\")\n",
    "obj<-SKAT_Null_Model(test_phenotypes$IIH_mod ~ data.matrix(FAM_Cov[,7:14]), out_type=\"C\")\n",
    "\n",
    "SSD.INFO<-Open_SSD(File.SSD, File.Info)\n",
    "out<-SKAT.SSD.All(SSD.INFO, obj)\n",
    "\n",
    "write.table(out$results,args[7])\n",
    "''')\n",
    "    with open(PYTHON_FILE,\"w\") as file:\n",
    "        file.write(\"import pandas as pd\\n\")\n",
    "        file.write(\"import sys\\n\")\n",
    "        file.write(\"import os\\n\")\n",
    "        file.write(\"df = pd.read_csv('\"+CSV_FILE+\"')\\n\")\n",
    "        file.write(\"out_file = '\" + BASE_PATH + \"/out_'+  str(sys.argv[1]) +'_\"+str(chrom)+\".csv'\\n\")\n",
    "        file.write(\"line = df.iloc[int(sys.argv[1]),]\\n\")\n",
    "        file.write(\"os.system('/cs/cbio/netanel/lib/R4/bin/Rscript \" + R_FILE + \" ' + line.bed + ' ' + line.bim + ' ' + line.fam + ' ' + line.SetID + ' ' + line.SSD + ' ' + line.info+ ' ' + out_file)\\n\")\n",
    "        file.write(\"\\n\")\n",
    "    with open(BASH_FILE,\"w\") as file:\n",
    "        file.write(f\"python {PYTHON_FILE} $SLURM_ARRAY_TASK_ID\\n\")\n",
    "    !chmod 744 {BASH_FILE}\n",
    "    !echo {BASH_FILE}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_0.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_0.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_1.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_1.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_2.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_2.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_3.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_3.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_4.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_4.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_5.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_5.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_6.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_6.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_7.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_7.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_8.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_8.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_9.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_9.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_10.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_10.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_11.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_11.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_12.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_12.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_13.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_13.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_14.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_14.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_15.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_15.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_16.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_16.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_17.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_17.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_18.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_18.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_19.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_19.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_20.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_20.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_21.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_21.sh running\n",
      "sleep 5m\n",
      "sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_22.sh\"\n",
      "sleep 1m\n",
      "echo /cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_22.sh running\n",
      "sleep 5m\n"
     ]
    }
   ],
   "source": [
    "for chrom in range(0,23):\n",
    "    BASH_FILE = f\"/cs/labs/michall/roeizucker/hackton_meno/scripts/create_out_{chrom}.sh\"\n",
    "    print(f'sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"{BASH_FILE}\"')\n",
    "    print(\"sleep 1m\")\n",
    "    print(f\"echo {BASH_FILE} running\")\n",
    "    print(\"sleep 5m\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create output files from SSD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "CHROM = 2\n",
    "# SET_ID_PATH = \"/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me\"\n",
    "CSV_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_SSD_{CHROM}.csv\"\n",
    "BASE_PATH = \"/cs/labs/michall/roeizucker/IIH\"\n",
    "CURR_BED = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{CHROM}.bed\"\n",
    "CURR_BIM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{CHROM}.bim\"\n",
    "CURR_FAM = f\"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{CHROM}.fam\"\n",
    "\n",
    "R_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_out_{CHROM}.R\"\n",
    "PYTHON_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_out_{CHROM}.py\"\n",
    "# CSV_FILE = \"/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/scripts/create_SSD.csv\"\n",
    "BASH_FILE = f\"/cs/labs/michall/roeizucker/IIH/scripts/create_out_{CHROM}.sh\"\n",
    "with open(R_FILE,\"w\") as file:\n",
    "    file.write('''require(SKAT)\n",
    "args = commandArgs(trailingOnly=TRUE)\n",
    "File.Bed<-args[1]\n",
    "File.Bim<-args[2]\n",
    "File.Fam<-args[3]\n",
    "File.SetID<-args[4]\n",
    "File.SSD<-args[5]\n",
    "File.Info<-args[6]\n",
    "\n",
    "SSD.INFO<-Open_SSD(File.SSD, File.Info)\n",
    "\n",
    "test_phenotypes = read.csv(file = \"/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/phens.csv\")\n",
    "obj<-SKAT_Null_Model(test_phenotypes$IIH_mod ~ 1, out_type=\"C\")\n",
    "SSD.INFO<-Open_SSD(File.SSD, File.Info)\n",
    "out<-SKAT.SSD.All(SSD.INFO, obj)\n",
    "print(args[7])\n",
    "write.table(out$results,args[7])\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(PYTHON_FILE,\"w\") as file:\n",
    "    file.write(\"import pandas as pd\\n\")\n",
    "    file.write(\"import sys\\n\")\n",
    "    file.write(\"import os\\n\")\n",
    "    file.write(\"df = pd.read_csv('\"+CSV_FILE+\"')\\n\")\n",
    "    file.write(\"out_file = '\" + BASE_PATH + \"/out_'+  str(sys.argv[1]) +'_\"+str(CHROM)+\".csv'\\n\")\n",
    "    file.write(\"line = df.iloc[int(sys.argv[1]),]\\n\")\n",
    "    file.write(\"os.system('/cs/cbio/netanel/lib/R4/bin/Rscript \" + R_FILE + \" ' + line.bed + ' ' + line.bim + ' ' + line.fam + ' ' + line.SetID + ' ' + line.SSD + ' ' + line.info+ ' ' + out_file)\\n\")\n",
    "    file.write(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(BASH_FILE,\"w\") as file:\n",
    "    file.write(f\"python {PYTHON_FILE} $SLURM_ARRAY_TASK_ID\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "!chmod 744 {BASH_FILE}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/labs/michall/roeizucker/IIH/scripts/create_out_2.sh\r\n"
     ]
    }
   ],
   "source": [
    "!ls \"/cs/labs/michall/roeizucker/IIH/scripts/create_out_2.sh\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "create_out_2.py  create_out.py\tcreate_SSD_2.csv  create_SSD.py  test.R\r\n",
      "create_out_2.R\t create_out.R\tcreate_SSD_3.R\t  create_SSD.R\r\n",
      "create_out_2.sh  create_out.sh\tcreate_SSD.csv\t  create_SSD.sh\r\n"
     ]
    }
   ],
   "source": [
    "!ls /cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/scripts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64948\n"
     ]
    }
   ],
   "source": [
    "print(counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create new phenotype file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>p31</th>\n",
       "      <th>p21001_i0</th>\n",
       "      <th>p36_i0</th>\n",
       "      <th>p4080_i0_a0</th>\n",
       "      <th>p4080_i0_a1</th>\n",
       "      <th>p2443_i0</th>\n",
       "      <th>p21022</th>\n",
       "      <th>p30890_i0</th>\n",
       "      <th>p21762_i2</th>\n",
       "      <th>...</th>\n",
       "      <th>p21662_i3</th>\n",
       "      <th>p5326_i0</th>\n",
       "      <th>p5327_i0</th>\n",
       "      <th>p5327_i1</th>\n",
       "      <th>p5326_i1</th>\n",
       "      <th>p20004_i0</th>\n",
       "      <th>p41270</th>\n",
       "      <th>p41280_a0</th>\n",
       "      <th>p34</th>\n",
       "      <th>IIH_mod</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eid2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3830686</th>\n",
       "      <td>3830686</td>\n",
       "      <td>Female</td>\n",
       "      <td>32.8933</td>\n",
       "      <td>1039.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>No</td>\n",
       "      <td>54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hysterectomy|bilateral oophorectomy|ovarian cy...</td>\n",
       "      <td>A09.9 Gastroenteritis and colitis of unspecifi...</td>\n",
       "      <td>2014-10-17</td>\n",
       "      <td>1955</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5450636</th>\n",
       "      <td>5450636</td>\n",
       "      <td>Female</td>\n",
       "      <td>53.5396</td>\n",
       "      <td>3158.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>No</td>\n",
       "      <td>56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reduction or fixation of bone fracture|dilatat...</td>\n",
       "      <td>E03.9 Hypothyroidism, unspecified|E66.8 Other ...</td>\n",
       "      <td>2011-06-13</td>\n",
       "      <td>1953</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5021098</th>\n",
       "      <td>5021098</td>\n",
       "      <td>Female</td>\n",
       "      <td>22.6718</td>\n",
       "      <td>1049.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>No</td>\n",
       "      <td>64</td>\n",
       "      <td>27.2</td>\n",
       "      <td>Staff 158935</td>\n",
       "      <td>...</td>\n",
       "      <td>2334.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D12.8 Rectum|H47.1 Papilloedema, unspecified|I...</td>\n",
       "      <td>2004-09-27</td>\n",
       "      <td>1944</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2401715</th>\n",
       "      <td>2401715</td>\n",
       "      <td>Male</td>\n",
       "      <td>25.7589</td>\n",
       "      <td>2815.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>No</td>\n",
       "      <td>58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E03.9 Hypothyroidism, unspecified|H47.1 Papill...</td>\n",
       "      <td>2020-11-05</td>\n",
       "      <td>1951</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4181381</th>\n",
       "      <td>4181381</td>\n",
       "      <td>Male</td>\n",
       "      <td>32.6190</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>No</td>\n",
       "      <td>61</td>\n",
       "      <td>56.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vasectomy</td>\n",
       "      <td>A09.9 Gastroenteritis and colitis of unspecifi...</td>\n",
       "      <td>2013-04-10</td>\n",
       "      <td>1948</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3827597</th>\n",
       "      <td>3827597</td>\n",
       "      <td>Female</td>\n",
       "      <td>27.5597</td>\n",
       "      <td>1046.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>No</td>\n",
       "      <td>68</td>\n",
       "      <td>72.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bunion/hallus valgus surgery|shoulder surgery</td>\n",
       "      <td>A49.0 Staphylococcal infection, unspecified|B9...</td>\n",
       "      <td>2015-10-11</td>\n",
       "      <td>1940</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4122899</th>\n",
       "      <td>4122899</td>\n",
       "      <td>Female</td>\n",
       "      <td>23.5102</td>\n",
       "      <td>156.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>No</td>\n",
       "      <td>42</td>\n",
       "      <td>55.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tonsillectomy +/- adenoids</td>\n",
       "      <td>F32.9 Depressive episode, unspecified|G93.2 Be...</td>\n",
       "      <td>2014-02-17</td>\n",
       "      <td>1967</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4407910</th>\n",
       "      <td>4407910</td>\n",
       "      <td>Male</td>\n",
       "      <td>33.2713</td>\n",
       "      <td>162.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>No</td>\n",
       "      <td>46</td>\n",
       "      <td>63.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male circumcision</td>\n",
       "      <td>D12.8 Rectum|D50.9 Iron deficiency anaemia, un...</td>\n",
       "      <td>2015-01-12</td>\n",
       "      <td>1962</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5281791</th>\n",
       "      <td>5281791</td>\n",
       "      <td>Female</td>\n",
       "      <td>23.6336</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>No</td>\n",
       "      <td>50</td>\n",
       "      <td>48.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heart surgery|upper limb surgery</td>\n",
       "      <td>E03.9 Hypothyroidism, unspecified|G43.9 Migrai...</td>\n",
       "      <td>2007-02-27</td>\n",
       "      <td>1957</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4360102</th>\n",
       "      <td>4360102</td>\n",
       "      <td>Female</td>\n",
       "      <td>29.9556</td>\n",
       "      <td>1046.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>134.0</td>\n",
       "      <td>No</td>\n",
       "      <td>65</td>\n",
       "      <td>23.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hysterectomy|bunion/hallus valgus surgery|musc...</td>\n",
       "      <td>B96.8 Other specified bacterial agents as the ...</td>\n",
       "      <td>2008-03-05</td>\n",
       "      <td>1943</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             eid     p31  p21001_i0  p36_i0  p4080_i0_a0  p4080_i0_a1  \\\n",
       "eid2                                                                    \n",
       "3830686  3830686  Female    32.8933  1039.0        150.0        145.0   \n",
       "5450636  5450636  Female    53.5396  3158.0        123.0        128.0   \n",
       "5021098  5021098  Female    22.6718  1049.0        152.0        127.0   \n",
       "2401715  2401715    Male    25.7589  2815.0        123.0        118.0   \n",
       "4181381  4181381    Male    32.6190  1053.0        183.0        181.0   \n",
       "...          ...     ...        ...     ...          ...          ...   \n",
       "3827597  3827597  Female    27.5597  1046.0        153.0        149.0   \n",
       "4122899  4122899  Female    23.5102   156.0        130.0        125.0   \n",
       "4407910  4407910    Male    33.2713   162.0        146.0        139.0   \n",
       "5281791  5281791  Female    23.6336  1053.0        127.0        119.0   \n",
       "4360102  4360102  Female    29.9556  1046.0        141.0        134.0   \n",
       "\n",
       "        p2443_i0  p21022  p30890_i0     p21762_i2  ...  p21662_i3  p5326_i0  \\\n",
       "eid2                                               ...                        \n",
       "3830686       No      54        NaN           NaN  ...        NaN       NaN   \n",
       "5450636       No      56        NaN           NaN  ...        NaN       NaN   \n",
       "5021098       No      64       27.2  Staff 158935  ...     2334.0       NaN   \n",
       "2401715       No      58        NaN           NaN  ...        NaN       NaN   \n",
       "4181381       No      61       56.9           NaN  ...        NaN       NaN   \n",
       "...          ...     ...        ...           ...  ...        ...       ...   \n",
       "3827597       No      68       72.4           NaN  ...        NaN       NaN   \n",
       "4122899       No      42       55.2           NaN  ...        NaN       NaN   \n",
       "4407910       No      46       63.1           NaN  ...        NaN       NaN   \n",
       "5281791       No      50       48.5           NaN  ...        NaN       NaN   \n",
       "4360102       No      65       23.0           NaN  ...        NaN       NaN   \n",
       "\n",
       "        p5327_i0 p5327_i1 p5326_i1  \\\n",
       "eid2                                 \n",
       "3830686      NaN      NaN      NaN   \n",
       "5450636      NaN      NaN      NaN   \n",
       "5021098      NaN      NaN      NaN   \n",
       "2401715      NaN      NaN      NaN   \n",
       "4181381      NaN      NaN      NaN   \n",
       "...          ...      ...      ...   \n",
       "3827597      NaN      NaN      NaN   \n",
       "4122899      NaN      NaN      NaN   \n",
       "4407910      NaN      NaN      NaN   \n",
       "5281791      NaN      NaN      NaN   \n",
       "4360102      NaN      NaN      NaN   \n",
       "\n",
       "                                                 p20004_i0  \\\n",
       "eid2                                                         \n",
       "3830686  hysterectomy|bilateral oophorectomy|ovarian cy...   \n",
       "5450636  reduction or fixation of bone fracture|dilatat...   \n",
       "5021098                                                NaN   \n",
       "2401715                                                NaN   \n",
       "4181381                                          vasectomy   \n",
       "...                                                    ...   \n",
       "3827597      bunion/hallus valgus surgery|shoulder surgery   \n",
       "4122899                         tonsillectomy +/- adenoids   \n",
       "4407910                                  male circumcision   \n",
       "5281791                   heart surgery|upper limb surgery   \n",
       "4360102  hysterectomy|bunion/hallus valgus surgery|musc...   \n",
       "\n",
       "                                                    p41270   p41280_a0   p34  \\\n",
       "eid2                                                                           \n",
       "3830686  A09.9 Gastroenteritis and colitis of unspecifi...  2014-10-17  1955   \n",
       "5450636  E03.9 Hypothyroidism, unspecified|E66.8 Other ...  2011-06-13  1953   \n",
       "5021098  D12.8 Rectum|H47.1 Papilloedema, unspecified|I...  2004-09-27  1944   \n",
       "2401715  E03.9 Hypothyroidism, unspecified|H47.1 Papill...  2020-11-05  1951   \n",
       "4181381  A09.9 Gastroenteritis and colitis of unspecifi...  2013-04-10  1948   \n",
       "...                                                    ...         ...   ...   \n",
       "3827597  A49.0 Staphylococcal infection, unspecified|B9...  2015-10-11  1940   \n",
       "4122899  F32.9 Depressive episode, unspecified|G93.2 Be...  2014-02-17  1967   \n",
       "4407910  D12.8 Rectum|D50.9 Iron deficiency anaemia, un...  2015-01-12  1962   \n",
       "5281791  E03.9 Hypothyroidism, unspecified|G43.9 Migrai...  2007-02-27  1957   \n",
       "4360102  B96.8 Other specified bacterial agents as the ...  2008-03-05  1943   \n",
       "\n",
       "         IIH_mod  \n",
       "eid2              \n",
       "3830686        1  \n",
       "5450636        1  \n",
       "5021098        1  \n",
       "2401715        1  \n",
       "4181381        1  \n",
       "...          ...  \n",
       "3827597        1  \n",
       "4122899        1  \n",
       "4407910        1  \n",
       "5281791        1  \n",
       "4360102        1  \n",
       "\n",
       "[267 rows x 21 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "participants_dataframe = pd.read_csv(FILE_CONTAINING_PARTICIPANTS)\n",
    "participants_dataframe\n",
    "participants_dataframe[\"IIH_mod\"] = 1\n",
    "participants_dataframe[\"eid2\"] = participants_dataframe[\"eid\"]\n",
    "participants_dataframe.set_index(\"eid2\",inplace=True)\n",
    "participants_dataframe\n",
    "# print(len(old_phenotype_dataframe))\n",
    "# old_phenotype_dataframe = pd.read_csv(OLD_PHENOTYPE_FILE)\n",
    "# old_phenotype_dataframe[\"eid2\"] = old_phenotype_dataframe[\"eid\"]\n",
    "# old_phenotype_dataframe.set_index(\"eid2\",inplace=True)\n",
    "# old_phenotype_dataframe[\"IIH_mod\"] = participants_dataframe[\"IIH_mod\"]\n",
    "# old_phenotype_dataframe[\"IIH_mod\"] = old_phenotype_dataframe[\"IIH_mod\"].fillna(0)\n",
    "# print(len(old_phenotype_dataframe))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_index</th>\n",
       "      <th>eid</th>\n",
       "      <th>F329</th>\n",
       "      <th>F329_F</th>\n",
       "      <th>F329_M</th>\n",
       "      <th>H26</th>\n",
       "      <th>H26_F</th>\n",
       "      <th>H26_M</th>\n",
       "      <th>H269</th>\n",
       "      <th>H269_F</th>\n",
       "      <th>...</th>\n",
       "      <th>batch_UKBiLEVEAX_b10</th>\n",
       "      <th>batch_UKBiLEVEAX_b11</th>\n",
       "      <th>batch_UKBiLEVEAX_b3</th>\n",
       "      <th>batch_UKBiLEVEAX_b4</th>\n",
       "      <th>batch_UKBiLEVEAX_b5</th>\n",
       "      <th>batch_UKBiLEVEAX_b6</th>\n",
       "      <th>batch_UKBiLEVEAX_b7</th>\n",
       "      <th>batch_UKBiLEVEAX_b8</th>\n",
       "      <th>batch_UKBiLEVEAX_b9</th>\n",
       "      <th>IIH_mod</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eid2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1609417</th>\n",
       "      <td>3593</td>\n",
       "      <td>1609417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3463715</th>\n",
       "      <td>7575</td>\n",
       "      <td>3463715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4555602</th>\n",
       "      <td>18424</td>\n",
       "      <td>4555602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3900040</th>\n",
       "      <td>24792</td>\n",
       "      <td>3900040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1219769</th>\n",
       "      <td>27479</td>\n",
       "      <td>1219769</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000071</th>\n",
       "      <td>473788</td>\n",
       "      <td>1000071</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3075338</th>\n",
       "      <td>478422</td>\n",
       "      <td>3075338</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2308410</th>\n",
       "      <td>495843</td>\n",
       "      <td>2308410</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2454182</th>\n",
       "      <td>497210</td>\n",
       "      <td>2454182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5862881</th>\n",
       "      <td>497506</td>\n",
       "      <td>5862881</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows  188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_index      eid  F329  F329_F  F329_M  H26  H26_F  H26_M  H269  \\\n",
       "eid2                                                                            \n",
       "1609417          3593  1609417   NaN     NaN     NaN  NaN    NaN    NaN   NaN   \n",
       "3463715          7575  3463715   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "4555602         18424  4555602   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "3900040         24792  3900040   NaN     NaN     NaN  NaN    NaN    NaN   NaN   \n",
       "1219769         27479  1219769   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "...               ...      ...   ...     ...     ...  ...    ...    ...   ...   \n",
       "1000071        473788  1000071   0.0     NaN     0.0  1.0    NaN    1.0   1.0   \n",
       "3075338        478422  3075338   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "2308410        495843  2308410   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "2454182        497210  2454182   0.0     NaN     0.0  1.0    NaN    1.0   1.0   \n",
       "5862881        497506  5862881   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "\n",
       "         H269_F  ...  batch_UKBiLEVEAX_b10  batch_UKBiLEVEAX_b11  \\\n",
       "eid2             ...                                               \n",
       "1609417     NaN  ...                     0                     0   \n",
       "3463715     NaN  ...                     0                     0   \n",
       "4555602     NaN  ...                     0                     0   \n",
       "3900040     NaN  ...                     0                     0   \n",
       "1219769     NaN  ...                     0                     0   \n",
       "...         ...  ...                   ...                   ...   \n",
       "1000071     NaN  ...                     0                     1   \n",
       "3075338     NaN  ...                     0                     0   \n",
       "2308410     NaN  ...                     0                     0   \n",
       "2454182     NaN  ...                     0                     0   \n",
       "5862881     NaN  ...                     0                     0   \n",
       "\n",
       "         batch_UKBiLEVEAX_b3  batch_UKBiLEVEAX_b4  batch_UKBiLEVEAX_b5  \\\n",
       "eid2                                                                     \n",
       "1609417                    0                    0                    0   \n",
       "3463715                    0                    0                    0   \n",
       "4555602                    0                    0                    0   \n",
       "3900040                    0                    0                    0   \n",
       "1219769                    0                    0                    0   \n",
       "...                      ...                  ...                  ...   \n",
       "1000071                    0                    0                    0   \n",
       "3075338                    0                    0                    0   \n",
       "2308410                    0                    0                    0   \n",
       "2454182                    0                    0                    0   \n",
       "5862881                    0                    0                    0   \n",
       "\n",
       "         batch_UKBiLEVEAX_b6  batch_UKBiLEVEAX_b7  batch_UKBiLEVEAX_b8  \\\n",
       "eid2                                                                     \n",
       "1609417                    0                    0                    0   \n",
       "3463715                    0                    0                    0   \n",
       "4555602                    0                    0                    0   \n",
       "3900040                    0                    0                    0   \n",
       "1219769                    0                    0                    1   \n",
       "...                      ...                  ...                  ...   \n",
       "1000071                    0                    0                    0   \n",
       "3075338                    0                    0                    0   \n",
       "2308410                    0                    0                    0   \n",
       "2454182                    0                    0                    0   \n",
       "5862881                    0                    0                    0   \n",
       "\n",
       "         batch_UKBiLEVEAX_b9  IIH_mod  \n",
       "eid2                                   \n",
       "1609417                    0      1.0  \n",
       "3463715                    0      1.0  \n",
       "4555602                    0      1.0  \n",
       "3900040                    0      1.0  \n",
       "1219769                    0      1.0  \n",
       "...                      ...      ...  \n",
       "1000071                    0      1.0  \n",
       "3075338                    0      1.0  \n",
       "2308410                    0      1.0  \n",
       "2454182                    0      1.0  \n",
       "5862881                    0      1.0  \n",
       "\n",
       "[70 rows x 188 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_phenotype_dataframe[(old_phenotype_dataframe[\"IIH_mod\"] == 1) & (old_phenotype_dataframe[\"sex\"] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sort new dataframe - change, only for presentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEST_BIM_PATH = \"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered.bim\"\n",
    "TEST_BED_PATH = \"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered.bed\"\n",
    "TEST_FAM_PATH = \"/cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered.fam\"\n",
    "PLINK_PATH = \"/cs/usr/nadavb/third_party/plink2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pdp.read_plink1_bin(TEST_BED_PATH,TEST_BIM_PATH,TEST_FAM_PATH,verbose=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "eid_vals_in_files =  list(a['iid'].values)\n",
    "eid_vals_in_files_set = set(eid_vals_in_files)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_phenotype_dataframe = old_phenotype_dataframe[old_phenotype_dataframe['eid'].isin(eid_vals_in_files)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274215\n",
      "274215\n"
     ]
    }
   ],
   "source": [
    "print(len(old_phenotype_dataframe))\n",
    "print(len(eid_vals_in_files))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cs/usr/roeizucker/my_python/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "old_phenotype_dataframe['sort_key'] = old_phenotype_dataframe['eid'].astype(\"str\").apply(lambda x: eid_vals_in_files.index(x) if x in eid_vals_in_files_set else float('inf'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_phenotype_dataframe = old_phenotype_dataframe.sort_values(by='sort_key')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_phenotype_dataframe[[\"eid\",\"IIH_mod\"]].to_csv(PHENOTYPE_FILE_NAME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/phens.csv\n"
     ]
    }
   ],
   "source": [
    "print(PHENOTYPE_FILE_NAME)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_index</th>\n",
       "      <th>eid</th>\n",
       "      <th>F329</th>\n",
       "      <th>F329_F</th>\n",
       "      <th>F329_M</th>\n",
       "      <th>H26</th>\n",
       "      <th>H26_F</th>\n",
       "      <th>H26_M</th>\n",
       "      <th>H269</th>\n",
       "      <th>H269_F</th>\n",
       "      <th>...</th>\n",
       "      <th>batch_UKBiLEVEAX_b11</th>\n",
       "      <th>batch_UKBiLEVEAX_b3</th>\n",
       "      <th>batch_UKBiLEVEAX_b4</th>\n",
       "      <th>batch_UKBiLEVEAX_b5</th>\n",
       "      <th>batch_UKBiLEVEAX_b6</th>\n",
       "      <th>batch_UKBiLEVEAX_b7</th>\n",
       "      <th>batch_UKBiLEVEAX_b8</th>\n",
       "      <th>batch_UKBiLEVEAX_b9</th>\n",
       "      <th>IIH_mod</th>\n",
       "      <th>sort_key</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eid2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1688461</th>\n",
       "      <td>411779</td>\n",
       "      <td>1688461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5444187</th>\n",
       "      <td>352393</td>\n",
       "      <td>5444187</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2676900</th>\n",
       "      <td>258733</td>\n",
       "      <td>2676900</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3709342</th>\n",
       "      <td>150866</td>\n",
       "      <td>3709342</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3769417</th>\n",
       "      <td>270220</td>\n",
       "      <td>3769417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1176921</th>\n",
       "      <td>320265</td>\n",
       "      <td>1176921</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>487364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1339882</th>\n",
       "      <td>263982</td>\n",
       "      <td>1339882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>487365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1470962</th>\n",
       "      <td>330791</td>\n",
       "      <td>1470962</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>487366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1193917</th>\n",
       "      <td>233981</td>\n",
       "      <td>1193917</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>487381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196108</th>\n",
       "      <td>129496</td>\n",
       "      <td>1196108</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>487388</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>332689 rows  189 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_index      eid  F329  F329_F  F329_M  H26  H26_F  H26_M  H269  \\\n",
       "eid2                                                                            \n",
       "1688461        411779  1688461   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "5444187        352393  5444187   NaN     NaN     NaN  NaN    NaN    NaN   NaN   \n",
       "2676900        258733  2676900   0.0     0.0     NaN  0.0    0.0    NaN   0.0   \n",
       "3709342        150866  3709342   0.0     0.0     NaN  0.0    0.0    NaN   0.0   \n",
       "3769417        270220  3769417   0.0     0.0     NaN  0.0    0.0    NaN   0.0   \n",
       "...               ...      ...   ...     ...     ...  ...    ...    ...   ...   \n",
       "1176921        320265  1176921   0.0     0.0     NaN  1.0    1.0    NaN   1.0   \n",
       "1339882        263982  1339882   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "1470962        330791  1470962   0.0     NaN     0.0  0.0    NaN    0.0   0.0   \n",
       "1193917        233981  1193917   0.0     0.0     NaN  1.0    1.0    NaN   1.0   \n",
       "1196108        129496  1196108   0.0     0.0     NaN  0.0    0.0    NaN   0.0   \n",
       "\n",
       "         H269_F  ...  batch_UKBiLEVEAX_b11  batch_UKBiLEVEAX_b3  \\\n",
       "eid2             ...                                              \n",
       "1688461     NaN  ...                     0                    0   \n",
       "5444187     NaN  ...                     0                    0   \n",
       "2676900     0.0  ...                     0                    0   \n",
       "3709342     0.0  ...                     0                    0   \n",
       "3769417     0.0  ...                     0                    0   \n",
       "...         ...  ...                   ...                  ...   \n",
       "1176921     1.0  ...                     1                    0   \n",
       "1339882     NaN  ...                     1                    0   \n",
       "1470962     NaN  ...                     1                    0   \n",
       "1193917     1.0  ...                     1                    0   \n",
       "1196108     0.0  ...                     1                    0   \n",
       "\n",
       "         batch_UKBiLEVEAX_b4  batch_UKBiLEVEAX_b5  batch_UKBiLEVEAX_b6  \\\n",
       "eid2                                                                     \n",
       "1688461                    0                    0                    0   \n",
       "5444187                    0                    0                    0   \n",
       "2676900                    0                    0                    0   \n",
       "3709342                    0                    0                    0   \n",
       "3769417                    0                    0                    0   \n",
       "...                      ...                  ...                  ...   \n",
       "1176921                    0                    0                    0   \n",
       "1339882                    0                    0                    0   \n",
       "1470962                    0                    0                    0   \n",
       "1193917                    0                    0                    0   \n",
       "1196108                    0                    0                    0   \n",
       "\n",
       "         batch_UKBiLEVEAX_b7  batch_UKBiLEVEAX_b8  batch_UKBiLEVEAX_b9  \\\n",
       "eid2                                                                     \n",
       "1688461                    0                    0                    0   \n",
       "5444187                    0                    0                    0   \n",
       "2676900                    0                    0                    0   \n",
       "3709342                    0                    0                    0   \n",
       "3769417                    0                    0                    0   \n",
       "...                      ...                  ...                  ...   \n",
       "1176921                    0                    0                    0   \n",
       "1339882                    0                    0                    0   \n",
       "1470962                    0                    0                    0   \n",
       "1193917                    0                    0                    0   \n",
       "1196108                    0                    0                    0   \n",
       "\n",
       "         IIH_mod  sort_key  \n",
       "eid2                        \n",
       "1688461      0.0         0  \n",
       "5444187      0.0         1  \n",
       "2676900      0.0         2  \n",
       "3709342      0.0         3  \n",
       "3769417      0.0         4  \n",
       "...          ...       ...  \n",
       "1176921      0.0    487364  \n",
       "1339882      0.0    487365  \n",
       "1470962      0.0    487366  \n",
       "1193917      0.0    487381  \n",
       "1196108      0.0    487388  \n",
       "\n",
       "[332689 rows x 189 columns]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_phenotype_dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create compatible Plink and phenotype files\n",
    "for SKAT to work the phenotype and genotype files need to be the same length, and have the same order. Participants that appear in only one dataset need to be removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE_PLINK_FORMAT_PATH = \"/cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch\"\n",
    "PLINK_PATH = \"/cs/usr/nadavb/third_party/plink2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a2.3LM 64-bit Intel (24 Jan 2020)   www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_1.log.\n",
      "Options in effect:\n",
      "  --keep-fam /cs/labs/michall/roeizucker/SKAT_experiments/I10_eid_notNA.txt\n",
      "  --make-bed\n",
      "  --out /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_1\n",
      "  --pgen /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bed\n",
      "  --psam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.fam\n",
      "  --pvar /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bim\n",
      "\n",
      "Start time: Wed Feb  1 22:31:12 2023\n",
      "128748 MiB RAM detected; reserving 64374 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "487409 samples (264362 females, 223033 males, 14 ambiguous; 487409 founders)\n",
      "loaded from\n",
      "/cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.fam.\n",
      "64451 variants loaded from\n",
      "/cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bim.\n",
      "Note: No phenotype data present.\n",
      "--keep-fam: 274215 samples remaining.\n",
      "274215 samples (148870 females, 125345 males; 274215 founders) remaining after\n",
      "main filters.\n",
      "Writing /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_1.fam ...\n",
      "done.\n",
      "Writing /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_1.bim ...\n",
      "done.\n",
      "Writing /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_1.bed ...\n",
      "0%"
     ]
    }
   ],
   "source": [
    "for i in range(1,2):\n",
    "    bim_path = BASE_PLINK_FORMAT_PATH + str(i) + \".bim\"\n",
    "    fam_path = BASE_PLINK_FORMAT_PATH + str(i) + \".fam\"\n",
    "    bed_path = BASE_PLINK_FORMAT_PATH + str(i) + \".bed\"\n",
    "    ! {PLINK_PATH} --bim {bim_path} --fam {fam_path} --bed {bed_path} --keep-fam /cs/labs/michall/roeizucker/SKAT_experiments/I10_eid_notNA.txt --make-bed --out /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{i}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! {PLINK_PATH} --bim {TEST_BIM_PATH} --fam {TEST_FAM_PATH} --bed {TEST_BED_PATH} --keep-fam /cs/labs/michall/roeizucker/SKAT_experiments/I10_eid_notNA.txt --make-bed --out /cs/labs/michall/roeizucker/SKAT_experiments/mydata_filtered_{i}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create phenotype files\n",
    "Phenotype files must be the same order and lenght as the genotype files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASIC_PHENOTYPE_FILE = \"/cs/usr/roeizucker/Desktop/temp/SKAT_delete_me/phens.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cs/usr/roeizucker/my_python/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3170: DtypeWarning: Columns (114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,254,256) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "basic_phenotype_df = pd.read_csv(BASIC_PHENOTYPE_FILE) # we will use this DF to determine order and length of phenotypes\n",
    "other_phenotype_df = pd.read_csv(\"/cs/labs/michall/roeizucker/ukb_files/menapause_with_outcomes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_phenotype_df[\"eid\"] = other_phenotype_df[\"Participant ID\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_phenotype_df[\"eid2\"] = other_phenotype_df[\"eid\"]\n",
    "other_phenotype_df = other_phenotype_df[other_phenotype_df['eid2'].isin(list(basic_phenotype_df[\"eid2\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "274215"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(basic_phenotype_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "basic_phenotype_df.set_index(['eid2'], inplace=True)\n",
    "other_phenotype_df.set_index(['eid2'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_phenotype_df[\"osteoporosis\"] = other_phenotype_df[\"M81.9 Osteoporosis, unspecified_Label\"]\n",
    "other_phenotype_df[\"fatty_liver\"] = other_phenotype_df[\"K76.0 Fatty (change of) liver, not elsewhere classified_Label\"]\n",
    "basic_phenotype_df[\"osteoporosis\"] = other_phenotype_df[\"osteoporosis\"]\n",
    "basic_phenotype_df[\"fatty_liver\"] = other_phenotype_df[\"fatty_liver\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "basic_phenotype_df[basic_phenotype_df[\"fatty_liver\"] == -1] = 0\n",
    "basic_phenotype_df[basic_phenotype_df[\"osteoporosis\"] == -1] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "basic_phenotype_df[\"osteoporosis\"] = basic_phenotype_df[\"osteoporosis\"].fillna(0)\n",
    "basic_phenotype_df[\"fatty_liver\"] = basic_phenotype_df[\"fatty_liver\"].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [],
   "source": [
    "basic_phenotype_df.to_csv(\"/cs/labs/michall/roeizucker/hackton_meno/labels_1.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cs/labs/michall/roeizucker/SKAT_experiments/SetID\n"
     ]
    }
   ],
   "source": [
    "print(PROJECT_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# find files for a specific gene\n",
    "import os, glob\n",
    "path = \"/cs/labs/michall/roeizucker/IIH\"\n",
    "for filename in glob.glob(os.path.join(path, '*.SetID')):\n",
    "    with open(os.path.join(os.getcwd(), filename), 'r') as f: # open in readonly mode\n",
    "        val = f.read()\n",
    "        if \"Q9UNW9\" in val:\n",
    "            print(filename)\n",
    "#             print(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find files for a specific gene\n",
    "import os, glob\n",
    "path = \"~\"\n",
    "for filename in glob.glob(os.path.join(path, '*.csv')):\n",
    "    with open(os.path.join(os.getcwd(), filename), 'r') as f: # open in readonly mode\n",
    "        if \"Q9UN\" in f.read():\n",
    "            print(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr1 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch2.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch2.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch2.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr2 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch3.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch3.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch3.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr3 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch4.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch4.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch4.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr4 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch5.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch5.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch5.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr5 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch6.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch6.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch6.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr6 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch7.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch7.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch7.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr7 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch8.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch8.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch8.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr8 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr9 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch10.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch10.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch10.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr10 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch11.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch11.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch11.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr11 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch12.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch12.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch12.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr12 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch13.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch13.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch13.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr13 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch14.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch14.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch14.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr14 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch15.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch15.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch15.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr15 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch16.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch16.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch16.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr16 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch17.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch17.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch17.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr17 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch18.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch18.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch18.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr18 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch19.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch19.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch19.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr19 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch20.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch20.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch20.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr20 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch21.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch21.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch21.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr21 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch22.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch22.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch22.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr22 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n"
     ]
    }
   ],
   "source": [
    "# create GWAS phenotype file:\n",
    "# f'sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"{BASH_FILE}\"')\n",
    "for i in range(1,23):\n",
    "    inner_command = (f\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{i}.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{i}.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{i}.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/amos/continuous_run/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr{i} --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\")\n",
    "    print(f'sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"{inner_command}\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch1.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr1 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch2.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch2.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch2.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr2 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch3.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch3.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch3.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr3 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch4.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch4.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch4.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr4 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch5.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch5.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch5.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr5 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch6.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch6.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch6.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr6 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch7.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch7.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch7.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr7 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch8.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch8.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch8.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr8 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch9.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr9 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch10.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch10.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch10.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr10 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch11.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch11.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch11.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr11 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch12.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch12.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch12.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr12 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch13.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch13.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch13.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr13 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch14.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch14.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch14.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr14 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch15.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch15.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch15.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr15 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch16.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch16.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch16.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr16 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch17.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch17.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch17.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr17 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch18.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch18.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch18.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr18 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch19.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch19.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch19.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr19 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch20.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch20.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch20.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr20 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch21.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch21.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch21.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr21 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n",
      "sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch22.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch22.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch22.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr22 --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\"\n"
     ]
    }
   ],
   "source": [
    "# small covariates /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt \n",
    "# create GWAS phenotype file:\n",
    "# f'sbatch --array=0-150 --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"{BASH_FILE}\"')\n",
    "for i in range(1,23):\n",
    "    inner_command = (f\"/cs/usr/nadavb/third_party/plink2 --bed /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{i}.bed --bim /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{i}.bim --fam /cs/labs/michall/roeizucker/plink_results/reduced_snps2/small_ch{i}.fam --pheno /cs/labs/michall/roeizucker/IIH/GWAS/phenotypes/iih_mod.txt --covar /cs/labs/michall/roeizucker/IIH/GWAS/covariates.txt  --out  /cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr{i} --1 --glm hide-covar --mac 20 --covar-variance-standardize --freq --threads 10 --memory 10000\")\n",
    "    print(f'sbatch --mem=15g -c1 --time=1-0 --killable --requeue --wrap=\"{inner_command}\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# create smaller covariats file:\n",
    "path = '/cs/labs/michall/roeizucker/10krun/runs/0:11'\n",
    "PHEN_PATH = path + \"/ukbb_dataset.csv\"\n",
    "GWAS_PATH = \"/cs/labs/michall/roeizucker/IIH/GWAS\"\n",
    "cancer_dataset = pd.read_csv(PHEN_PATH)\n",
    "# FAM_Cov[,7:14]\n",
    "\n",
    "covariate_cols = list(cancer_dataset)[15:22]\n",
    "covariate_values = cancer_dataset[['eid', 'eid'] + [covariate_col for covariate_col in covariate_cols if \\\n",
    "        covariate_col != 'const']].dropna()\n",
    "covariate_values.to_csv(f\"{GWAS_PATH}/covariates.txt\", header = False, index = False, sep = '\\t')\n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine plink results\n",
    "all_dfs = []\n",
    "for i in range(1,23):\n",
    "    df = pd.read_csv(f\"/cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_chr{i}.PHENO1.glm.logistic\", sep=\"\\t\")\n",
    "    all_dfs.append(df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine small covariates plink results\n",
    "all_small_covar_dfs = []\n",
    "for i in range(1,23):\n",
    "    df = pd.read_csv(f\"/cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_small_cov_chr{i}.PHENO1.glm.logistic\", sep=\"\\t\")\n",
    "    all_dfs.append(df)\n",
    "\n",
    "# save file\n",
    "small_cover_result = pd.concat(all_dfs)\n",
    "small_cover_result = small_cover_result.reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = pd.concat(all_dfs)\n",
    "result = result.reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>OBS_CT</th>\n",
       "      <th>OR</th>\n",
       "      <th>LOG(OR)_SE</th>\n",
       "      <th>Z_STAT</th>\n",
       "      <th>P</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>69428</td>\n",
       "      <td>rs140739101</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>185796.0</td>\n",
       "      <td>0.810917</td>\n",
       "      <td>0.711121</td>\n",
       "      <td>-0.294731</td>\n",
       "      <td>0.768199</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>rs568226429</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>215183.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>69496</td>\n",
       "      <td>rs150690004</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>215172.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>69521</td>\n",
       "      <td>rs553724620</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>214929.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>rs2531267</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>215160.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571731</th>\n",
       "      <td>6852</td>\n",
       "      <td>22</td>\n",
       "      <td>51183430</td>\n",
       "      <td>rs547756808</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274214.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571732</th>\n",
       "      <td>6853</td>\n",
       "      <td>22</td>\n",
       "      <td>51183575</td>\n",
       "      <td>rs531270968</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274101.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571733</th>\n",
       "      <td>6854</td>\n",
       "      <td>22</td>\n",
       "      <td>51183576</td>\n",
       "      <td>rs201653264</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>273408.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571734</th>\n",
       "      <td>6855</td>\n",
       "      <td>22</td>\n",
       "      <td>51183608</td>\n",
       "      <td>rs571073375</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274206.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571735</th>\n",
       "      <td>6856</td>\n",
       "      <td>22</td>\n",
       "      <td>51216386</td>\n",
       "      <td>rs75855559</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274175.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>571736 rows  14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index  #CHROM       POS           ID REF ALT A1 TEST    OBS_CT  \\\n",
       "0           0       1     69428  rs140739101   G   T  G  ADD  185796.0   \n",
       "1           1       1     69487  rs568226429   A   G  A  ADD  215183.0   \n",
       "2           2       1     69496  rs150690004   A   G  A  ADD  215172.0   \n",
       "3           3       1     69521  rs553724620   A   T  A  ADD  214929.0   \n",
       "4           4       1     69569    rs2531267   C   T  C  ADD  215160.0   \n",
       "...       ...     ...       ...          ...  ..  .. ..  ...       ...   \n",
       "571731   6852      22  51183430  rs547756808   G   C  G  ADD  274214.0   \n",
       "571732   6853      22  51183575  rs531270968   A   C  A  ADD  274101.0   \n",
       "571733   6854      22  51183576  rs201653264   A   G  A  ADD  273408.0   \n",
       "571734   6855      22  51183608  rs571073375   C   G  C  ADD  274206.0   \n",
       "571735   6856      22  51216386   rs75855559   G   T  G  ADD  274175.0   \n",
       "\n",
       "              OR  LOG(OR)_SE    Z_STAT         P  GENE  \n",
       "0       0.810917    0.711121 -0.294731  0.768199  None  \n",
       "1            NaN         NaN       NaN       NaN  None  \n",
       "2            NaN         NaN       NaN       NaN  None  \n",
       "3            NaN         NaN       NaN       NaN  None  \n",
       "4            NaN         NaN       NaN       NaN  None  \n",
       "...          ...         ...       ...       ...   ...  \n",
       "571731       NaN         NaN       NaN       NaN  None  \n",
       "571732       NaN         NaN       NaN       NaN  None  \n",
       "571733       NaN         NaN       NaN       NaN  None  \n",
       "571734       NaN         NaN       NaN       NaN  None  \n",
       "571735       NaN         NaN       NaN       NaN  None  \n",
       "\n",
       "[571736 rows x 14 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.8679229562150228\n",
      "1.7358459124300456\n",
      "2.603768868645068\n",
      "3.471691824860091\n",
      "4.339614781075114\n",
      "5.207537737290136\n",
      "6.075460693505159\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-66-909b50f2edbc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0msmall_cover_result\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"GENE\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;31m# add_genes_for_snps(result,genes_csv)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m \u001b[0madd_genes_for_snps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msmall_cover_result\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgenes_csv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;31m# result.loc[3,\"GENE\"] = \"bob\" small_cover_result\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-66-909b50f2edbc>\u001b[0m in \u001b[0;36madd_genes_for_snps\u001b[0;34m(df, genes_csv)\u001b[0m\n\u001b[1;32m      8\u001b[0m         \u001b[0mcurr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mgenes_csv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"chr == '\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"#CHROM\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"' and cds_start <= \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"POS\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"and cds_end >= \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"POS\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcurr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"uniprot_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m             \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"GENE\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcurr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"uniprot_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;31m#         row[\"GENE\"] = list(curr[\"uniprot_id\"])[0]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m#         sig_genes.add(list(curr[\"uniprot_id\"])[0])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    668\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    669\u001b[0m         \u001b[0miloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"iloc\"\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 670\u001b[0;31m         \u001b[0miloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_with_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    671\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    672\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_setitem_with_indexer\u001b[0;34m(self, indexer, value)\u001b[0m\n\u001b[1;32m   1761\u001b[0m                 \u001b[0;31m# scalar value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1762\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mloc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0milocs\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1763\u001b[0;31m                     \u001b[0misetter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1764\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1765\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36misetter\u001b[0;34m(loc, v)\u001b[0m\n\u001b[1;32m   1691\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1692\u001b[0m                 \u001b[0;31m# reset the sliced object if unique\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1693\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iset_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mser\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1694\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1695\u001b[0m             \u001b[0;31m# we need an iterable, with a ndim of at least 1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_iset_item\u001b[0;34m(self, loc, value)\u001b[0m\n\u001b[1;32m   3099\u001b[0m         \u001b[0;31m#  but the behavior is the same as long as we pass broadcast=False\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3100\u001b[0m         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sanitize_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbroadcast\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3101\u001b[0;31m         \u001b[0mNDFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iset_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3102\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3103\u001b[0m         \u001b[0;31m# check if we are modifying a copy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_iset_item\u001b[0;34m(self, loc, value)\u001b[0m\n\u001b[1;32m   3567\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3568\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_iset_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3569\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3570\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3571\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36miset\u001b[0;34m(self, loc, value)\u001b[0m\n\u001b[1;32m   1094\u001b[0m             \u001b[0mblk_locs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblklocs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mval_locs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1095\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mblk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshould_store\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1096\u001b[0;31m                 \u001b[0mblk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblk_locs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue_getitem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval_locs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1097\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m                 \u001b[0munfit_mgr_locs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmgr_locs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_array\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mblk_locs\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/my_python/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mset\u001b[0;34m(self, locs, values)\u001b[0m\n\u001b[1;32m    329\u001b[0m         \u001b[0mcreate\u001b[0m \u001b[0ma\u001b[0m \u001b[0mnew\u001b[0m \u001b[0marray\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0malways\u001b[0m \u001b[0mcreates\u001b[0m \u001b[0ma\u001b[0m \u001b[0mnew\u001b[0m \u001b[0mBlock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m         \"\"\"\n\u001b[0;32m--> 331\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlocs\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    332\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    333\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# add gene to result:\n",
    "def add_genes_for_snps(df,genes_csv):\n",
    "    sig_genes = set()\n",
    "    for i in range(len(df)):\n",
    "        if i % 10000 == 0:\n",
    "            print(i/len(df) * 100)\n",
    "        row = df.loc[i]\n",
    "        curr = (genes_csv.query(\"chr == '\" + str(row[\"#CHROM\"]) + \"' and cds_start <= \" + str(row[\"POS\"]) + \"and cds_end >= \" + str(row[\"POS\"])))\n",
    "        if len(list(curr[\"uniprot_id\"]))> 0:\n",
    "            df.loc[i,\"GENE\"] = list(curr[\"uniprot_id\"])[0]\n",
    "#         row[\"GENE\"] = list(curr[\"uniprot_id\"])[0]\n",
    "#         sig_genes.add(list(curr[\"uniprot_id\"])[0])\n",
    "    return sig_genes\n",
    "genes_csv = pd.read_csv(GENE_FILE_PATH)\n",
    "small_cover_result[\"GENE\"] = None\n",
    "# add_genes_for_snps(result,genes_csv)\n",
    "add_genes_for_snps(small_cover_result,genes_csv)\n",
    "\n",
    "# result.loc[3,\"GENE\"] = \"bob\" small_cover_result\n",
    "# print(result.loc[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv(\"/cs/labs/michall/roeizucker/IIH/GWAS/results/iih_mod_gwas.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "small_cover_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>OBS_CT</th>\n",
       "      <th>OR</th>\n",
       "      <th>LOG(OR)_SE</th>\n",
       "      <th>Z_STAT</th>\n",
       "      <th>P</th>\n",
       "      <th>GENE</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>69428</td>\n",
       "      <td>rs140739101</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>185796.0</td>\n",
       "      <td>0.810917</td>\n",
       "      <td>0.711121</td>\n",
       "      <td>-0.294731</td>\n",
       "      <td>0.768199</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>69428</td>\n",
       "      <td>rs140739101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>rs568226429</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>215183.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>rs568226429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>69496</td>\n",
       "      <td>rs150690004</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>215172.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>69496</td>\n",
       "      <td>rs150690004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>69521</td>\n",
       "      <td>rs553724620</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>214929.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>69521</td>\n",
       "      <td>rs553724620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>rs2531267</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>215160.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>rs2531267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571731</th>\n",
       "      <td>6852</td>\n",
       "      <td>22</td>\n",
       "      <td>51183430</td>\n",
       "      <td>rs547756808</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274214.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>22</td>\n",
       "      <td>51183430</td>\n",
       "      <td>rs547756808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571732</th>\n",
       "      <td>6853</td>\n",
       "      <td>22</td>\n",
       "      <td>51183575</td>\n",
       "      <td>rs531270968</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274101.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>22</td>\n",
       "      <td>51183575</td>\n",
       "      <td>rs531270968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571733</th>\n",
       "      <td>6854</td>\n",
       "      <td>22</td>\n",
       "      <td>51183576</td>\n",
       "      <td>rs201653264</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>273408.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>22</td>\n",
       "      <td>51183576</td>\n",
       "      <td>rs201653264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571734</th>\n",
       "      <td>6855</td>\n",
       "      <td>22</td>\n",
       "      <td>51183608</td>\n",
       "      <td>rs571073375</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274206.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>22</td>\n",
       "      <td>51183608</td>\n",
       "      <td>rs571073375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571735</th>\n",
       "      <td>6856</td>\n",
       "      <td>22</td>\n",
       "      <td>51216386</td>\n",
       "      <td>rs75855559</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>274175.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>22</td>\n",
       "      <td>51216386</td>\n",
       "      <td>rs75855559</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>571736 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index  #CHROM       POS           ID REF ALT A1 TEST    OBS_CT  \\\n",
       "0           0       1     69428  rs140739101   G   T  G  ADD  185796.0   \n",
       "1           1       1     69487  rs568226429   A   G  A  ADD  215183.0   \n",
       "2           2       1     69496  rs150690004   A   G  A  ADD  215172.0   \n",
       "3           3       1     69521  rs553724620   A   T  A  ADD  214929.0   \n",
       "4           4       1     69569    rs2531267   C   T  C  ADD  215160.0   \n",
       "...       ...     ...       ...          ...  ..  .. ..  ...       ...   \n",
       "571731   6852      22  51183430  rs547756808   G   C  G  ADD  274214.0   \n",
       "571732   6853      22  51183575  rs531270968   A   C  A  ADD  274101.0   \n",
       "571733   6854      22  51183576  rs201653264   A   G  A  ADD  273408.0   \n",
       "571734   6855      22  51183608  rs571073375   C   G  C  ADD  274206.0   \n",
       "571735   6856      22  51216386   rs75855559   G   T  G  ADD  274175.0   \n",
       "\n",
       "              OR  LOG(OR)_SE    Z_STAT         P  GENE  CHR        BP  \\\n",
       "0       0.810917    0.711121 -0.294731  0.768199  None    1     69428   \n",
       "1            NaN         NaN       NaN       NaN  None    1     69487   \n",
       "2            NaN         NaN       NaN       NaN  None    1     69496   \n",
       "3            NaN         NaN       NaN       NaN  None    1     69521   \n",
       "4            NaN         NaN       NaN       NaN  None    1     69569   \n",
       "...          ...         ...       ...       ...   ...  ...       ...   \n",
       "571731       NaN         NaN       NaN       NaN  None   22  51183430   \n",
       "571732       NaN         NaN       NaN       NaN  None   22  51183575   \n",
       "571733       NaN         NaN       NaN       NaN  None   22  51183576   \n",
       "571734       NaN         NaN       NaN       NaN  None   22  51183608   \n",
       "571735       NaN         NaN       NaN       NaN  None   22  51216386   \n",
       "\n",
       "                SNP  \n",
       "0       rs140739101  \n",
       "1       rs568226429  \n",
       "2       rs150690004  \n",
       "3       rs553724620  \n",
       "4         rs2531267  \n",
       "...             ...  \n",
       "571731  rs547756808  \n",
       "571732  rs531270968  \n",
       "571733  rs201653264  \n",
       "571734  rs571073375  \n",
       "571735   rs75855559  \n",
       "\n",
       "[571736 rows x 17 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import dash_bio\n",
    "\n",
    "# df = pd.read_csv('https://raw.githubusercontent.com/plotly/dash-bio-docs-files/master/manhattan_data.csv')\n",
    "\n",
    "result[\"CHR\"] = result[\"#CHROM\"]\n",
    "result[\"BP\"] = result[\"POS\"]\n",
    "result[\"SNP\"] = result[\"ID\"]\n",
    "\n",
    "dash_bio.ManhattanPlot(\n",
    "    dataframe=result,\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
